---
title: "Case study report for `r params$trait_name`"
date: "`r format(Sys.time(), '%d %B %Y')`"
output:
  html_document:
    df_print: paged
    toc: true
    number_sections: true
    toc_depth: 4
    toc_float:
      collapsed: false
      smooth_scroll: true
params:
  main: main
  main_lit: main_lit
  trait_name: trait_name
  lit_space_file: lit_space_file
---

```{r setup I, eval=TRUE, echo=F}
library(cowplot)
library(flextable)
library(officer)
library(readxl)
library(tidyr)
library(tidyverse)
library(TwoSampleMR)
library(vroom)
knitr::opts_chunk$set(eval = TRUE, include = T, message = F, warning = F, comment="")
```

```{r include=F}







source("../../helper_functions.R")
source("explore_MR-EvE_app/functions.R")
source("mr_related_functions.R")


set_flextable_defaults(big.mark = " ", 
  font.size = 8, theme_fun = theme_vanilla,
  background.color = "white")

```

```{r echo=F}
# odd thing I have to do because `params` list object is acting weird
params2<-params
params2$main -> main 
params2$main_lit -> main_lit
params2$trait_name -> trait_name
params2$lit_space_file -> lit_space_file
```



```{r ,  include=F, eval = F, echo=F}

#### HOW TO generate a case study report: â‚¬
# run one of this in console

rmarkdown::render("01_MR_related/extra_mr_for_case_studies.Rmd", 
  params = list(
    trait_name = "Cardiotrophin-1",
    main = "prot-a-710",
    main_lit = "prot-a-710",
    lit_space_file = "lit_terms_cardiotrophin1.csv"
))


rmarkdown::render("01_MR_related/extra_mr_for_case_studies.Rmd", 
                  params = list(
                    trait_name = "Age at menopause (last menstrual period)",
                    main = "ukb-b-17422",
                    main_lit = "ukb-b-17422",
                    lit_space_file = "lit_terms_menopause.csv"
                  ))


rmarkdown::render("01_MR_related/extra_mr_for_case_studies.Rmd", 
                  params  = list(
                    trait_name = "IGF-1",
                    main = "ukb-d-30770_irnt",
                    main_lit = "ukb-d-30770_irnt",
                    lit_space_file = "lit_terms_IGF1.csv"
                  ),
                   output_file = paste0("case_study_report_", gsub(" ", "_", trait_name, ".html"))
)

rmarkdown::render("01_MR_related/extra_mr_for_case_studies.Rmd", 
                  params = list(
                    trait_name = "Childhood body size",
                    main = "ukb-a-34",
                    main_lit = "ieu-a-1096",
                    lit_space_file = "lit_terms_childhood_obesity.csv"
                  ))




#main <- "prot-a-67"
#main_lit <- "prot-a-67"
#trait_name <- extract_instruments(main) %>% split_exposure() %>% pull(exposure) %>% unique()
#
#main <- "prot-b-38"
#main_lit <- "prot-b-38"
#trait_name <- extract_instruments(main) %>% split_exposure() %>% pull(exposure) %>% unique()
#
#main <- "prot-a-366"
#main_lit <- "prot-a-366"
#trait_name <- extract_instruments(main) %>% split_exposure() %>% pull(exposure) %>% unique()
#
```

**Report for: `r trait_name`**

# Main MR results

`r trait_name` was identified in MR-EvE search as having evidence of effect on breast cancer and validated with IVW MR method in BCAC 2017 and BCAC 2020 breast cancer outcomes: 


```{r echo =F, fig.height=3, fig.width=7}
path <- c("../results/mr_evidence_outputs/mr_subtypes/all_data_with_sens_filters.xlsx")
mr_df<- path %>% 
  excel_sheets() %>% 
  set_names() %>% 
  map(read_excel, path = path)

mr_results<-tibble()
for (i in names(mr_df)){
  tmp<- mr_df[[i]] %>% filter(id.exposure == main)
  mr_results <- bind_rows(mr_results, tmp)
}


 mr_results <- mr_results %>% 
   tidyr::separate(col = exposure, into = c("exposure", "tmp"), sep = "\\(" ) %>% 
   mutate(outcome = case_when(outcome =="Breast cancer BCAC 2020" ~ "BCAC 2020",
                             outcome =="Breast cancer (Combined Oncoarray; iCOGS; GWAS meta analysis)" ~ "BCAC 2017",
                             outcome =="ER+ Breast cancer (Combined Oncoarray; iCOGS; GWAS meta analysis)" ~ "ER+",
                             outcome =="ER- Breast cancer (Combined Oncoarray; iCOGS; GWAS meta analysis)" ~ "ER-",
                                           outcome =="LuminalA ER+PR+HER-"    ~ "Luminal A",   
                                           outcome =="LuminalB1 ER+PR+HER-"   ~ "Luminal B1",  
                                           outcome =="LuminalB2 ER+PR+HER+" ~ "Luminal B2" ,
                                           outcome =="HER2-enriched ER-PR-HER+"  ~ "HER2-enriched" ,
                                           outcome =="TNBC ER-PR-HER-"  ~ "TNBC" )) %>% 
   mutate(outcome = factor(outcome,  levels = c("BCAC 2017", "ER+", "ER-","BCAC 2020",
                                     "Luminal A", "Luminal B1", "Luminal B2", 
                                      "HER2-enriched", "TNBC" ))) %>% 
    mutate(outcome = factor(outcome, levels = rev(levels(outcome)))) %>% 
    mutate(effect_direction = ifelse(or_lci95 > 1 & or_uci95 >= 1, 'ok',
                              ifelse(or_lci95 < 1 & or_uci95 <= 1, 'ok', 'overlaps null'))) 

 
pal<-(c(unname(yarrr::piratepal("pony"))))
pal[6:7]<-c('darkgrey', "#FFEA5E")
p<-ggplot(mr_results, aes(y=outcome, x=or, colour=outcome, shape = effect_direction)) +
  geom_errorbarh(aes(xmin=or_lci95, xmax=or_uci95), height=.3) +
  geom_point(size=2)+
  scale_color_manual(values=pal)+
  scale_shape_manual(values=c(16,1))+
  geom_vline(xintercept=1, linetype='longdash') +
  geom_text(aes(label=OR_CI),hjust=-0.1, vjust=-0.6, size =3, color = '#333232')+
  theme_minimal_vgrid(10, rel_small = 1) +
  scale_y_discrete(position = "right")+
  facet_wrap( exposure ~ . , ncol=1)+
  labs(color = "", y = "Outcome", x = "Odds ratio" )+
  theme(legend.position = "none", plot.title.position  = "plot")
p

```

```{r echo =F}
 
if (mr_results[mr_results$outcome == "BCAC 2017",]$effect_direction == 'ok') {all =T} else{all =F}
if (mr_results[mr_results$outcome == "ER+",]$effect_direction == 'ok') {pos =T} else{pos =F}
if (mr_results[mr_results$outcome == "ER-",]$effect_direction == 'ok') {neg =T} else{neg =F}


mr_results_disp <- mr_results %>% select(exposure, outcome,  "IVW OR" = OR_CI, pval, nsnp,
                                    "Egger intercept < 0.05" = `egger_intercept_less_than_0.05`, 
                                    "Heterogeneity Qpval < 0.05" = `heterogeneity_Q_pval_less_than_0.05`) %>% 
                                mutate(pval = format(pval, digits=2))


ft<-flextable(mr_results_disp)
ft<-width(ft, j = c(1,2,3), width=1.5) 

ft
```


# MR-EvE-mined mediators

```{r,  include=F}

med_conf_table <- readxl::read_excel("../results/mr_evidence_outputs/med-conf-table_upd_ALL.xlsx", sheet = main) %>% 
     mutate(outcome = case_when(outcome.id =="ieu-a-1126" ~ "BCAC 2017",
                             outcome.id =="ieu-a-1127" ~ "ER+",
                             outcome.id =="ieu-a-1128" ~ "ER-"))

meds_table <- med_conf_table %>% 
  filter(type == 'mediator') %>% 
  select('exposure.trait', 
         'outcome',
         'outcome.id', 
         'med.id',
         'med.trait',
         'r1.OR_CI_val', 
         'r3.OR_CI_val') %>% 
  rename('exposure' = 'exposure.trait', 
         'mediator.id' = 'med.id',
         'mediator' = 'med.trait',
         "exposure -> mediator \n (step1)" = 'r1.OR_CI_val', 
         'mediator -> outcome \n (step2)' = 'r3.OR_CI_val') 


unique_meds <- length(unique(meds_table$mediator.id))


```

By querying MR-EvE data for `r trait_name` we identified `r unique_meds` potential mediators. This analysis was done only for BCAC 2017 outcomes (full / ER+ / ER- samples).

Below is a table of two-step MR results for each mediator, including _exposure -> mediator_ and	_mediator -> outcome_, and the total effect of _exposure -> outcome_. In the main analysis, `r trait_name` only had effect on `r if(all) {"total "}` , `r if(pos) {"ER+ "}` , `r if(neg) {"ER-"}` sample(s), so only them were used in MR-EvE mediators quesry.

## Two-step MR

```{r echo=F, results='asis'}
if (all){
  med <- meds_table %>% filter(outcome == "BCAC 2017") %>% pull( mediator.id) %>% length()
  cat(paste0("**Full sample**: ", med , " potential mediators from MR-EvE" ))}
```

```{r, echo=F}
if (all){
  ft<-flextable(meds_table %>% filter(outcome == "BCAC 2017") %>% select(-outcome.id) %>% arrange(desc(mediator.id)))
  ft<-width(ft, j = c(1), width=1.5)
  ft<-width(ft, j = c(4), width=3)
  ft<-width(ft, j = c(5,6), width=1.3)
  ft
}
```

```{r echo=F, results='asis'}
if (pos){
  med <- meds_table %>% filter(outcome == "ER+") %>% pull( mediator.id) %>% length()
  cat(paste0("**ER+ sample**: ", med , " potential mediators from MR-EvE" ))}
```

```{r echo =F}
if (pos) {
  ft<-flextable(meds_table %>% filter(outcome == "ER+") %>% select(-outcome.id) %>% arrange(desc(mediator.id)))
  ft<-width(ft, j = c(1), width=1.5)
  ft<-width(ft, j = c(4), width=3)
  ft<-width(ft, j = c(5,6), width=1.3)
  ft
  }
```


```{r echo=F, results='asis'}
if (neg){
  med <- meds_table %>% filter(outcome == "ER-") %>% pull( mediator.id) %>% length()
  cat(paste0("**ER- sample:** ", med , " potential mediators from MR-EvE" ))}
```

```{r echo =F}
if (neg) {
  ft<-flextable(meds_table %>% filter(outcome == "ER-")%>% select(-outcome.id) %>% arrange(desc(mediator.id)))
  ft<-width(ft, j = c(1), width=1.5)
  ft<-width(ft, j = c(4), width=3)
  ft<-width(ft, j = c(5,6), width=1.3)
  ft
}
```

## MVMR

```{r include=F, cache=T}

### this is MVMR (next section)

trait = unique(meds_table$exposure)
outcomes <- unique(meds_table$outcome.id) 

disrupted_by_all <- tibble()

for (bc_out in outcomes){

  interm_list<-med_conf_table %>%  filter(med.id %in% meds_table$mediator.id) %>% filter(outcome.id == bc_out) %>%  pull(med.id)
  
  df<-tibble()
  for ( interm in interm_list){
    tmp<- quick_mvmr(exp1 = main, exp2 = interm, out = bc_out) %>% 
      select(1:2, OR_CI, effect_direction, nsnp) %>%
      mutate(effect = "direct", interm = interm) 
    df<- bind_rows(df, tmp)
  }

  
  # list of interm where exp or/and interm overlaps null
  x<- df %>% filter(effect_direction == 'overlaps null') %>% pull(interm)
  
  # exposure disrupted by interm
  # returns disrupted exposure values
  disrupted_by <- df %>% filter(interm %in% x) %>% 
    filter(exposure == trait) %>% 
    mutate(is_disrupted = ifelse(effect_direction == 'overlaps null', T,F)) %>%
    arrange(exposure) %>% distinct()
  
  disrupted_by<- disrupted_by %>% 
    left_join(med_conf_table %>% filter(outcome.id == bc_out) %>%
                                 select( med.id,med.trait, r2.OR_CI_val), 
              by = c('interm' = 'med.id')) %>% 
    rename(disrupter = med.trait, OR_CI_direct = OR_CI, OR_CI_total = r2.OR_CI_val) %>% 
    select(exposure,  outcome, OR_CI_total, disrupter,  OR_CI_total, OR_CI_direct, is_disrupted, interm)
  
  disrupted_by_all <- bind_rows(disrupted_by_all, disrupted_by) 

  
}

disrupted_by_all<-
  disrupted_by_all %>%  rename("Total effect \n exposure -> outcome" = "OR_CI_total",
                             "Direct effect \n exposure -> outcome" = "OR_CI_direct",
                             "mediator" = 'disrupter') %>% arrange(desc(interm)) %>% 
          select(exposure, outcome, mediator, everything()) %>% select(-interm)


attenuate_full <- disrupted_by_all %>% filter(outcome == "Breast cancer ", is_disrupted == T) %>% pull(mediator) 
attenuate_pos <- disrupted_by_all %>% filter(outcome == "ER+ Breast cancer ", is_disrupted == T) %>% pull(mediator) 
attenuate_neg <- disrupted_by_all %>% filter(outcome == "ER- Breast cancer ", is_disrupted == T) %>% pull(mediator) 
```



```{r, echo=F, results= 'asis'}
if (all){
  disrupted_by_all_display<- disrupted_by_all %>% distinct() %>% filter(outcome == "Breast cancer ")
  cat(paste0(length(attenuate_full), " traits attenuate the effect of ", trait_name, " on **breast cancer (full sample)** (highlighted):"))
}
```

```{r, echo=F}
if (all){
  ft<-flextable(disrupted_by_all_display)
  ft<-width(ft, j = c(1,2,4,5), width=1.5)
  ft<-width(ft, j = 3, width=3)
  ft %>% bg(
    ., i = ~ is_disrupted == T, j = "mediator",
    bg = "#E9B4DF", part = "body")
}
```

```{r, echo=F, results='asis'}
if (pos){
  disrupted_by_all_display<- disrupted_by_all %>% distinct() %>% filter(outcome == "ER+ Breast cancer ")
  cat(paste0(length(attenuate_pos), " traits attenuate the effect of ", trait_name, " on **ER+ breast cancer** (highlighted):"))
}
```

```{r, echo=F}
if (pos){
  ft<-flextable(disrupted_by_all_display)
  ft<-width(ft, j = c(1,2,4,5), width=1.5)
  ft<-width(ft, j = 3, width=3)
  ft %>% bg(
    ., i = ~ is_disrupted == T, j = "mediator",
    bg = "#E9B4DF", part = "body")
}
```

```{r, echo=F, results='asis'}
if (neg){
  disrupted_by_all_display<- disrupted_by_all %>% distinct() %>% filter(outcome == "ER- Breast cancer ")
  cat(paste0(length(attenuate_neg), " traits attenuate the effect of ", trait_name, " on **ER- breast cancer** (highlighted):"))
  
}
```

```{r, echo=F}
if (neg){
  ft<-flextable(disrupted_by_all_display)
  ft<-width(ft, j = c(1,2,4,5), width=1.5)
  ft<-width(ft, j = 3, width=3)
  ft %>% bg(
    ., i = ~ is_disrupted == T, j = "mediator",
    bg = "#E9B4DF", part = "body")
}
```



# Literature-mined mediators



```{r, include =F}
lit_df<- read_tsv("../../02_literature_related/literature_outputs/traits_marked_for_lit_analysis.tsv")
lit_trait <-lit_df %>% filter(id == main_lit)


case_terms<- read_csv(paste0("../../02_literature_related/literature_outputs/sankey_terms_storage/", lit_space_file)) 
if (main == "prot-a-710") {
  case_terms <-case_terms %>% filter(value != 'cardiotrophin 1')
}

```


```{r  cache=TRUE, include=F, eval =T}
lit_terms_df<- case_terms %>% select(value, gwas.id, gwas.name) %>% drop_na() %>% distinct() 

for (i in 1:length(lit_terms_df$gwas.id)){
  y <- extract_instruments(lit_terms_df$gwas.id[i])
  if (is.null(y)){
    lit_terms_df$instruments[i] <- 0
  } else{
    
    if ( NA %in% y$eaf.exposure) {
      lit_terms_df$instruments[i] <- 0
    } else {
      lit_terms_df$instruments[i] <- dim(y)[1]
    }
  }
}


ids <- lit_terms_df %>% 
  filter(instruments > 0 ) %>% 
  filter(!gwas.id %in% c("prot-a-81", "prot-a-1102", "ukb-b-8587", "prot-a-2694", "prot-a-2363", "finn-b-E4_OBESITY") )%>% 
  pull(gwas.id) %>% unique()
```

```{r , include=F, eval =T}
# for reporting
lit_terms <- unique(case_terms$value) # all lit terms
unique_lit_terms<- unique(lit_terms_df$value) #  lit terms with gwas data
gwas_with_inst_total <-lit_terms_df %>% filter(instruments > 0) %>% pull(value)
gwas_with_inst_unique <-lit_terms_df %>% filter(instruments > 0) %>% pull(value) %>% unique()

traits_can_use <- lit_terms_df %>% filter(instruments > 0 ) %>% pull(gwas.name) %>% unique()

# terms with at least 1 inst 
#lit_terms_df %>% filter(nSNPs >= 1) %>% select(lit_term, gwas.name) %>% distinct() 
```

## Summary

`r trait_name` literature space contains: 

 - `r lit_trait$unique_triples` unique triples of terms

 - `r lit_trait$unique_pairs`  unique pairs of terms
 
## Literature overlap Sankey plot

```{r}
#library(networkD3)


### subset of sankey plot of lit overlap goes here


```

## Matching literature terms to GWAS traits

`r trait_name` and breast cancer literature space overlap identified: 
 
 - `r length(lit_terms)` unique literature terms
 
 - `r length(unique_lit_terms)` terms are available as GWAS traits in OpenGWAS (table below)
 
 - `r length(gwas_with_inst_total)` (`r length(gwas_with_inst_unique)` unique) traits have >=1 genome-wide significant SNPs (instruments for MR) (highlighted in the table)

```{r echo=F, eval =T}
lit_terms_df_disp <- lit_terms_df %>% rename("Literature term"="value",
                                        "Matching GWAS ID"="gwas.id" , 
                                        "Matching GWAS" = "gwas.name",
                                        "nSNPs p < 5x10e8" = "instruments") %>% 
                select("Literature term","Matching GWAS", "Matching GWAS ID", "nSNPs p < 5x10e8" )

ft<-flextable(lit_terms_df_disp)
ft<-width(ft, j = c(1), width=2)
ft<-width(ft, j = c(2), width=4)
ft<-width(ft, j = c(3), width=1.3)

ft %>% bg(
  ., i = ~ `nSNPs p < 5x10e8` > 0, 
  j = colnames(lit_terms_df_disp), 
  bg = "#FDB382", part = "body")
```


## Two-step MR


```{r , include=F, eval =T}
res_df_all<-tibble()
res_df_pos<-tibble()
res_df_neg<-tibble()

## do MR for all gwas from the list; for outputs that were set
if (all){ res_df_all <- bind_rows(lapply(ids, do_MR, 'all'))  }
if (pos){ res_df_pos <- bind_rows(lapply(ids, do_MR, 'ER+'))  }
if (neg){ res_df_neg <- bind_rows(lapply(ids, do_MR, 'ER-'))  }

redone_MR_full <- bind_rows(res_df_all,res_df_pos, res_df_neg)

# split into MR and sens and save
redone_MR <- redone_MR_full %>% 
              select(-starts_with('egger_intercept'), -starts_with('Q'))

##
redoneMR_tidy <- redone_MR %>%  
  tidyr::separate(col = outcome, into = c("outcome", "tmp"), sep = "\\(" ) %>% 
  filter(effect_direction != "overlaps null") %>% 
  filter(method %in% c('Inverse variance weighted', 'Wald ratio')) %>% arrange(outcome) %>% 
  select(exposure, id.exposure ,  outcome, id.outcome, OR_CI,  nsnp, effect_direction) 
```

Then we performed two-step MR for the traits with >=1 instruments:

```{r unordered_list1, echo=FALSE, results='asis'}
cat(paste('-', traits_can_use), sep = '\n')
```

Out of these traits, `r length(unique(redoneMR_tidy$exposure))` had evidence of effect on breast cancer (so could be considered further as mediators):

```{r unordered_list2, echo=FALSE, results='asis'}
cat(paste('-', unique(redoneMR_tidy$exposure)), sep = '\n')
```

**Mediator -> outcome (step2) MR results** for the traits with evidence of effect:

```{r echo=F, eval =T}
ft <- flextable(redoneMR_tidy %>% select(-id.outcome, -effect_direction))
ft <- width(ft, j = c(1,3,4), width=2)
ft
```

## Bidirectional MR 

We next performed bidirectional MR to to establish the direction of the effect between the intermediate and the risk factor and identify the traits that are affected by the risk factor (i.e. likely mediator relationship) [**Exposure -> mediator (step1)**] and the other way around (i.e. potential confounder relationship). 


```{r , echo=F, eval =T, cache=T}
step1_mr <- tibble()

for (interm in unique(redoneMR_tidy$id.exposure)){
  biMR <- bind_rows(
      quick_mr(exp = main, out = interm) %>%
        select(exposure,id.exposure,  outcome,id.outcome, beta_CI, effect_direction, nsnp) %>% mutate(effect = "term trait as mediator"),
      quick_mr(exp = interm, out = main) %>%
        select(exposure,id.exposure,  outcome,id.outcome,  beta_CI, effect_direction, nsnp) %>% mutate(effect = "term trait as confounder")
      )
  step1_mr <- bind_rows(step1_mr, biMR)
}

potential_mediators <- step1_mr %>% filter(effect_direction != 'overlaps null', effect == "term trait as mediator" ) %>% pull(outcome)

ft <- flextable(step1_mr)
ft <- width(ft, j = c(1,3,5,8), width=3)

border <- fp_border( width = 2)

ft %>%
  bg(., 
      i = ~ effect_direction != 'overlaps null', 
      j = colnames(step1_mr), 
      bg = "#7CC5A7", part = "body") %>% 
  hline(.,
        i = ~ effect == "term trait as confounder", border = fp_border( width = 2), part = "body")
  

```

Potential mediators: **`r paste0(potential_mediators, collapse=", ")`**


## MVMR

```{r , echo=F, cache =TRUE, eval =T}

outcomes <- unique(redoneMR_tidy$id.outcome)  
total_res <- tibble()


for (bc_out in outcomes){
  
  results_list <- list()
  
  interm_list <- redoneMR_tidy %>% 
              filter(exposure %in% potential_mediators) %>%  ### only of do those that pass two-step MR
              filter(id.outcome == bc_out) %>% 
              filter(!id.exposure %in% c('prot-a-2363', "prot-a-757")) %>% pull(id.exposure)
  
  for (interm in interm_list){
  
    x<-full_join(
        quick_mr(exp = main, out = bc_out) %>% 
          select(exposure, outcome, "total_OR_CI" = OR_CI, "total_ED" = effect_direction, "total_OR" = or) %>% 
          mutate(outcome = gsub("(Combined Oncoarray; iCOGS; GWAS meta analysis)", "", outcome, fixed = T)) ,# main to out
        quick_mr(exp = interm, out = bc_out) %>% 
          select(exposure, outcome, "total_OR_CI" = OR_CI, "total_ED" = effect_direction, "total_OR" = or) %>% 
          mutate(outcome = gsub("(Combined Oncoarray; iCOGS; GWAS meta analysis)", "", outcome, fixed = T)) ) %>% 
     full_join(.,
        quick_mvmr(exp1 = main, exp2 = interm, out = bc_out) %>%
          select(exposure, outcome, "direct_OR_CI" = OR_CI, "direct_ED" = effect_direction, "direct_OR" = or) ) %>%
      mutate(test = paste0(interm, " / ", bc_out) ) %>% 
      select(test, exposure, everything()) 
    
    results_list[[interm]] <- x
  }  
  
  all_res <- bind_rows(results_list)
  total_res <-bind_rows(total_res, all_res) 
  
}

#rm(total_res_out)
total_res_out<- transform(total_res,group =as.numeric(factor(test)))
total_res_out_dips<- total_res_out %>%
  select(group, everything(), -test) %>% 
  arrange(group) %>%
  select(-total_OR, -direct_OR) %>% 
  rename("#" = "group", 
         "total effect (MR)" = "total_OR_CI",  "direct effect (MVMR)" = "direct_OR_CI",
         "total effect direction" = "total_ED",  "direct effect direction" = "direct_ED") 
```




```{r echo =F, eval =T}
total_res_out_dips <- as_grouped_data(x = total_res_out_dips, groups = c("#"), columns = NULL)
ft<-flextable(total_res_out_dips)
ft<-width(ft, j = c(2), width=2.5)
ft<-width(ft, j = c(3:7), width=1.8)
ft %>% bg(
  ., i = ~ `direct effect direction` == 'overlaps null', 
  j = "direct effect direction", 
  bg = "#6CA3F8", part = "body")
```




























# (Optional): sensitivity analysis with female only data (if available)



## Bidirectional MR

```{r echo =F,  eval =F}
interm_list <- redoneMR_tidy %>% filter(id.outcome == bc_out) %>% filter(id.exposure !=  'prot-a-2363') %>% pull(id.exposure) %>% unique()
#instruments<-read_tsv("/Users/ny19205/OneDrive - University of Bristol/Documents - OneDrive/Mini-project2/01_Data/GWAS_tophits/igf_tophits.tsv")
instruments<-read_tsv("/Users/ny19205/OneDrive - University of Bristol/Documents - OneDrive/Mini-project2/01_Data/GWAS_tophits/early_bmi_adj_tophits.tsv")

out_df<-tibble()

for (i in interm_list){
  
  out <- extract_outcome_data(
    snps = instruments$SNP,
    outcome = i)

  harmonised<- harmonise_data(exposure_dat = instruments, 
                              outcome_dat = out)
  
  res_early <- TwoSampleMR::mr(harmonised, method_list = c('mr_ivw')) %>% 
    split_outcome() %>% 
    split_exposure() %>% 
    generate_odds_ratios() %>% 
    select(exposure, outcome, starts_with("or"))
  
  out_df<-bind_rows(out_df, res_early)
  
}

out_df %>% kable_it()

```


```{r echo =F,  eval =F}
# opposite dir

interm_list <- redoneMR_tidy %>% filter(id.outcome == bc_out) %>% filter(id.exposure !=  'prot-a-2363') %>% pull(id.exposure) %>% unique()

#gwas<- vroom::vroom("/Users/ny19205/OneDrive - University of Bristol/Documents - OneDrive/Mini-project2/01_Data/GWAS_results_tidy/igf_GWAS_tidy_outcome.txt.gz")
gwas<- vroom::vroom("/Users/ny19205/OneDrive - University of Bristol/Documents - OneDrive/Mini-project2/01_Data/GWAS_results_tidy/early_bmi_adj_GWAS_tidy_outcome.txt.gz")



out_df2<-tibble()

for (i in interm_list){
  
  instruments <- extract_instruments(i)
  
  out <- gwas %>% 
        filter(SNP %in%instruments$SNP)

  harmonised<- harmonise_data(exposure_dat = instruments, 
                              outcome_dat = out)
  
  res_early <- TwoSampleMR::mr(harmonised, method_list = c('mr_ivw')) %>% 
    split_outcome() %>% 
    split_exposure() %>% 
    generate_odds_ratios() %>% 
    select(exposure, outcome, starts_with("or"))
  
  out_df2<-bind_rows(out_df2, res_early)
  
}

out_df2 %>% kable_it()


```


## MVMR

```{r echo =F,  eval =F}


#id2_tophits_file  <- "igf_tophits.tsv"
#id2_gwas_file <- "igf_GWAS_tidy_outcome.txt.gz"
library(vroom)
id2_tophits_file  <-  vroom("/Users/ny19205/OneDrive - University of Bristol/Documents - OneDrive/Mini-project2/01_Data/GWAS_tophits/early_bmi_adj_tophits.tsv")

id2_gwas_file <- vroom("/Users/ny19205/OneDrive - University of Bristol/Documents - OneDrive/Mini-project2/01_Data/GWAS_results_tidy/early_bmi_adj_GWAS_tidy_outcome.txt.gz")


# to try: redoneMR_tidy

#igf1_meds <-c('ukb-b-16881', 'ukb-a-269','prot-b-20','ieu-b-4869','finn-b-E4_OBESITY','ieu-a-974')

med_list <- unique(redoneMR_tidy$id.exposure)

all_res <- tibble()

for (i in med_list){
  x1 <- mvmr_mixed_sources(id1 = i,
                   outcome.id = 'ieu-a-1126',
                   id2_tophits_file,
                   id2_gwas_file)
  
  x2 <- mvmr_mixed_sources(id1 = i,
                   outcome.id = 'ieu-a-1127',
                   id2_tophits_file,
                   id2_gwas_file)
  x3 <- mvmr_mixed_sources(id1 = i,
                   outcome.id = 'ieu-a-1128',
                   id2_tophits_file,
                   id2_gwas_file)
  
  
  all_res<- bind_rows(all_res, x1, x2, x3)
}

```
