---
title: "Process traits"
output: html_notebook
---

```{r message=F}
library(dplyr)
library(readr)
library(tidyr)
library(stringr)
library(tibble)
library(TwoSampleMR)

source("explore_MR-EvE_app/functions.R")
source("mr_related_functions.R")
```


```{r, message = FALSE, include=F}
rm(dat)
dat <-  read_tsv("../results/mr_evidence_outputs/tidy_traits_by_cat.tsv") %>%  # made in explore_mr_results.Rmd
        # update exposure categories 
        create_exposure_categories() %>%  add_exposure_labels() %>% filter(exposure_cat != 'other')

exp_help_names <- dat %>% # from the first chunk
  select('exposure_name'='exposure', 'id.exposure'='exposure.id', 'exposure_cat') %>% 
  distinct()

dat %>% select(exposure.id, exposure_cat) %>% distinct() %>%  count(exposure_cat, sort=T)
n<- dat %>% select(exposure.id, exposure_cat) %>% distinct() %>% count(exposure_cat) %>% summarise(across(n, sum)) %>% pull(n)# 905

table1_col1<- dat %>% select(exposure.id, exposure_cat) %>% distinct() %>%  count(exposure_cat, sort=T)


dat %>% select(exposure_cat, exposure.trait, exposure.id) %>% distinct() %>% filter(exposure_cat == 'Other biomarkers') %>% View()
dat %>% select(exposure_cat, exposure.trait, exposure.id) %>% distinct() %>% filter(exposure_cat == 'Alcohol') %>% View()


```





```{r}
# select traits that  have effect in 2/3 of main dataset in a subtype

all_cats<-tibble() # run once
cats <- dat %>% pull(exposure_cat) %>% unique()

for (i in cats){
  cat('\n')
  print(paste0("Doing ", i))
  
  input <- dat %>% filter(exposure_cat == i)
  
  # run for each category input -- 2/3 method
  u <- traits_for_follow_up(input)
  
  all_cats<-bind_rows(all_cats, u) %>% distinct()
}

all_cats %>% select(exposure_cat, trait, id) %>% distinct() %>% count(exposure_cat, sort=T)
table1_col2<- all_cats %>% select(exposure_cat, trait, id) %>% distinct() %>% count(exposure_cat, sort=T)

dim(all_cats) # 309

write_tsv(all_cats, "../results/mr_evidence_outputs/trait_for_followup.tsv")
```

```{r}
## summary MOE methods for that data that we keep

dat %>% 
  filter(exposure.id %in% all_cats$id) %>% 
  filter(chip %in% c('Meta', 'iCOG2017', 'OncArray')) %>% 
  select(exposure.id, outcome.id, mr.method) %>% 
  distinct() %>% 
  count(mr.method) %>% arrange(-n) %>% 
  mutate(pct =  round(n/sum(n)*100, 2)) 

```


#### Next go to `03sub_validate_mr.Rmd` to get validation results


```{r}
# load BCAC 2017 validation results and turn into wide format

# create wide form of results to view by subtype
redone_MR <- read_tsv("../results/mr_evidence_outputs/redone_MR_fulloutput.tsv") # created in 035
length(unique(redone_MR$id.exposure)) # 309

redone_wide <- redone_MR %>%
  select(id.outcome, id.exposure,exposure, exposure_name, effect_direction, method, exposure_cat) %>% 
  filter(method %in% c('Inverse variance weighted', 'Wald ratio') )%>% 
  mutate(effect_dir_binary = case_when(effect_direction == 'negative' ~ -1, 
                                       effect_direction == 'positive' ~  1,
                                       effect_direction == 'overlaps null' ~  0)) %>% 
  select(-effect_direction, -method) %>% 
  pivot_wider(names_from = id.outcome, values_from = effect_dir_binary)

```


```{r message =F}
# load BCAC 2020 validation results and turn into wide format

res_subtypes <- read_tsv("../results/mr_evidence_outputs/mr_subtypes/all_traits_MR_vs_BCAC2020.tsv") %>% 
                filter(outcome != "TNBC_BRCA1 ER-PR-HER-")

## category update 
res_subtypes <- res_subtypes %>% 
  rename(exposure.trait=exposure, exposure.id = id.exposure ) %>%
  create_exposure_categories() %>%  filter(exposure_cat != 'other') %>% 
  select(-exposure ) %>% 
  rename(id.exposure = exposure.id , exposure = exposure.trait) 


res_subtypes_wide<- res_subtypes %>% 
  select(outcome, id.exposure,exposure,  effect_direction, method) %>% 
  filter(method %in% c('Inverse variance weighted', 'Wald ratio') )%>% 
  mutate(effect_dir_binary = case_when(effect_direction == 'negative' ~ -1, 
                                       effect_direction == 'positive' ~  1,
                                       effect_direction == 'overlaps null' ~  0)) %>% 
  select(-effect_direction, -method) %>% 
  pivot_wider(names_from = outcome, values_from = effect_dir_binary)

```


```{r}

res_subtypes_for_count <- bind_rows(redone_MR,res_subtypes) %>% 
  select(exposure, id.exposure , outcome, OR_CI, effect_direction , nsnp, method, exposure_cat) %>% 
  filter(effect_direction != "overlaps null") %>% 
  filter(method %in% c('Inverse variance weighted', 'Wald ratio')) 

length(unique(res_subtypes$exposure_cat))
unique(res_subtypes_for_count$outcome)

counts_by_outcome <- list(
  total = res_subtypes_for_count %>% select(id.exposure, exposure_cat) %>% distinct() %>%  count(exposure_cat) %>% rename(total=n) %>% 
    column_to_rownames('exposure_cat') %>% t() %>% as_data_frame(),
  total2017 = res_subtypes_for_count %>% filter(grepl("Combined", outcome)) %>% select(id.exposure, exposure_cat) %>% distinct() %>%  count(exposure_cat) %>% rename(total2017=n) %>% 
    column_to_rownames('exposure_cat') %>% t() %>% as_data_frame(),
   all2017 = res_subtypes_for_count %>% filter(outcome =="Breast cancer (Combined Oncoarray; iCOGS; GWAS meta analysis)")%>% count(exposure_cat) %>% rename(all2017=n)%>%
    column_to_rownames('exposure_cat') %>% t() %>% as_data_frame(),
  er_pos = res_subtypes_for_count %>% filter(outcome =="ER+ Breast cancer (Combined Oncoarray; iCOGS; GWAS meta analysis)")%>% count(exposure_cat) %>% rename(er_pos=n)%>% 
    column_to_rownames('exposure_cat') %>% t() %>% as_data_frame(),
  er_neg = res_subtypes_for_count %>% filter(outcome =="ER- Breast cancer (Combined Oncoarray; iCOGS; GWAS meta analysis)")%>% count(exposure_cat) %>% rename(er_neg=n)%>% 
    column_to_rownames('exposure_cat') %>% t() %>% as_data_frame(),
  total2020 = res_subtypes_for_count %>% filter(!grepl("Combined", outcome)) %>% select(id.exposure, exposure_cat) %>% distinct() %>%  count(exposure_cat) %>% rename(total2020=n) %>% 
    column_to_rownames('exposure_cat') %>% t() %>% as_data_frame(),
  all2020 = res_subtypes_for_count %>% filter(outcome =="Breast cancer BCAC 2020")%>% count(exposure_cat) %>% rename(all2020=n)%>%
    column_to_rownames('exposure_cat') %>% t() %>% as_data_frame(),
  lumA = res_subtypes_for_count %>% filter(outcome =="LuminalA ER+PR+HER-")%>% count(exposure_cat) %>% rename(lumA=n)%>% 
    column_to_rownames('exposure_cat') %>% t() %>% as_data_frame(),
  lumB1 = res_subtypes_for_count %>% filter(outcome =="LuminalB1 ER+PR+HER-")%>% count(exposure_cat) %>% rename(lumB1=n)%>% 
    column_to_rownames('exposure_cat') %>% t() %>% as_data_frame(),
  lumB2 = res_subtypes_for_count %>% filter(outcome =="LuminalB2 ER+PR+HER+")%>% count(exposure_cat) %>% rename(lumB2=n)%>% 
    column_to_rownames('exposure_cat') %>% t() %>% as_data_frame(),
  her2 = res_subtypes_for_count %>% filter(outcome =="HER2-enriched ER-PR-HER+")%>% count(exposure_cat) %>% rename(her2=n)%>% 
    column_to_rownames('exposure_cat') %>% t() %>% as_data_frame(),
  tnbc = res_subtypes_for_count %>% filter(outcome =="TNBC ER-PR-HER-")%>% count(exposure_cat) %>% rename(tnbc=n)%>% 
    column_to_rownames('exposure_cat') %>% t() %>% as_data_frame())

counts_by_outcome_df <- bind_rows(counts_by_outcome) %>% t() %>% as.data.frame() %>% rownames_to_column('exposure_cat') 
colnames(counts_by_outcome_df) <- c("exposure_cat", names(counts_by_outcome)) 

preval_counts <- full_join(table1_col1%>% rename(total_preval=n),
                           table1_col2 %>% rename(total_consist=n))

counts_by_outcome_df<- left_join(preval_counts, counts_by_outcome_df) %>% column_to_rownames('exposure_cat')  %>% replace(is.na(.), 0) 

sums<- colSums(counts_by_outcome_df)
counts_by_outcome_df<- counts_by_outcome_df%>% bind_rows(.,sums)

table1 <- counts_by_outcome_df %>% 
    rownames_to_column('exposure_cat') %>% 
    mutate(exposure_cat = case_when(exposure_cat == "...13" ~"TOTAL", 
                                    exposure_cat == 'Antrophometric' ~ 'Anthropometric',
                                     exposure_cat == 'Drugs' ~ 'Medication',
                                    TRUE ~ exposure_cat) )%>% 
      mutate(exposure_cat = factor(exposure_cat, 
     levels= c( "Anthropometric", "Diet and supplements", "Physical activity", "Reproductive", 
               "Sleep", "Medication", "Alcohol", "Smoking", "Proteins", "Lipids", "Metabolites", "Other biomarkers", "TOTAL"))) %>% 
     arrange(exposure_cat) %>% 
    mutate(group = ifelse(exposure_cat %in% c("Proteins", "Lipids", "Metabolites", "Other biomarkers"), "Molecular traits", "Lifestyle traits")) %>% 
    mutate(group = ifelse(exposure_cat == "TOTAL", "TOTAL", group)) %>% 
    select(group, everything()) 
  

write_tsv(counts_by_outcome_df, "../results/mr_evidence_outputs/mr_subtypes/table1_counts_by_exposure_cat.tsv")
```

```{r}
library(flextable)
library(officer)
set_flextable_defaults(big.mark = " ", 
  font.size = 8, theme_fun = theme_vanilla,
  background.color = "white")

table1_disp<- transform(table1,group =factor(group))
ft<-flextable(table1_disp) %>% 
   align(align = "center", part = "all", j = c(3:16)) %>% 
  merge_v(j = ~group) %>% 
  width(., j = c(1,2), width=2) %>% 
 # width(., j = c(3), width=2) %>% 
  vline(.,j = c(1,2,3,4,5,9), border = fp_border( width = 2), part = "all") %>% 
  vline(.,j = c(6,7,8,10:15), border = fp_border( width = 1), part = "all") %>% 
  add_header_row(., 
                 values = c("",'Trait category',	
                            'Traits in MR-EvE app',	
                            'Traits with consistent effect *',
                            'Traits validated in either BCAC',
                            'Traits validated in \n BCAC 2017 (Supplementary Table 1)',
                            'Traits validated in \n BCAC 2020 (Supplementary Table 2)'),
                 colwidths = c(1, 1, 1, 1,1, 4, 7)) %>% 
  set_header_labels(ft,
    values = list(
            exposure_cat = "",
            group = " ",
            "total_preval" = 'Total',
            "total_consist" = 'Total',
            "total"   = 'Total',
            "total2017" = 'Total',
            "all2017" = 'Full',
            "er_pos" = "ER+",
            "er_neg"= "ER-",
            "total2020"  = 'Total',  
            "all2020" = 'Full',
            "lumA" = "LumA",
            "lumB1"  = "LumB1",
            "lumB2" = "LumB2",
            "her2"  = "HER2",
            "tnbc"  = "TNBC" ))
ft

```



```{r}
#merge subtypes and validation 
merged <- left_join(redone_wide, 
                    res_subtypes_wide) %>% 
          select(-exposure_name) %>% 
          left_join(exp_help_names) %>% 
          mutate(label = paste0("(", id.exposure, ") ", exposure_name )) %>% 
          select(exposure_cat, label, exposure_name, everything()) 


write_tsv(merged, "../results/mr_evidence_outputs/trait_manual_ivw_subtypes_merged.tsv")
```




```{r}
library(eulerr)

s4 <- list(`All` = redone_wide %>% filter(`ieu-a-1126` != 0) %>% pull(id.exposure) %>% sort(),
           `ER+` = redone_wide %>% filter(`ieu-a-1127` != 0) %>% pull(id.exposure)%>% sort(),
           `ER-` = redone_wide %>% filter(`ieu-a-1128` != 0) %>% pull(id.exposure)%>% sort())
plot(euler(s4, shape = "ellipse"), quantities = TRUE)

```

```{r}
s4 <- list(`All` = merged %>% filter(`ieu-a-1126` != 0) %>% pull(id.exposure) %>% sort(),
           `ER+` = merged %>% filter(`ieu-a-1127` != 0) %>% pull(id.exposure)%>% sort(),
           `ER-` = merged %>% filter(`ieu-a-1128` != 0) %>% pull(id.exposure)%>% sort(),
           
           `Breast cancer BCAC 2020` = merged %>% filter(`Breast cancer BCAC 2020` != 0) %>% pull(id.exposure)%>% sort(),
           `LuminalA ER+PR+HER`      = merged %>% filter(`LuminalA ER+PR+HER-` != 0) %>% pull(id.exposure)%>% sort(),
           `LuminalB1 ER+PR+HER-`    = merged %>% filter(`LuminalB1 ER+PR+HER-` != 0) %>% pull(id.exposure)%>% sort(),
           `HER2-enriched ER-PR-HER+` = merged %>% filter(`HER2-enriched ER-PR-HER+` != 0) %>% pull(id.exposure)%>% sort(),
           `TNBC ER-PR-HER-`         = merged %>% filter(`TNBC ER-PR-HER-` != 0) %>% pull(id.exposure)%>% sort(),
           `LuminalB2 ER+PR+HER+`     = merged %>% filter(`LuminalB2 ER+PR+HER+` != 0) %>% pull(id.exposure)%>% sort())


s4 <- list(
           `BCAC 2020 (n=176)` = merged %>% filter(`Breast cancer BCAC 2020` != 0 | 
                                           `LuminalA ER+PR+HER-`     != 0 |  
                                            `LuminalB1 ER+PR+HER-`    != 0 | 
                                            `HER2-enriched ER-PR-HER+`!= 0 | 
                                            `TNBC ER-PR-HER-`         != 0 | 
                                            `LuminalB2 ER+PR+HER+`    != 0 ) %>% pull(id.exposure) %>% sort(),
            `BCAC 2017 (n=168)` = merged %>% filter(`ieu-a-1128` != 0 | `ieu-a-1126` != 0 |`ieu-a-1127` != 0) %>% pull(id.exposure)%>% sort(),
            `passed MTC (n=105)` = passed_pairs %>% filter(mtc == "*") %>% pull(id.exposure) %>% unique() ,
            `Total (n=309)` = merged %>% pull(id.exposure)%>% sort())

plot(euler(s4, shape = "ellipse"), quantities = TRUE)
```







### Multiple testing

```{r}
# create wide form of results to view by subtype
redone_MR <- read_tsv("../results/mr_evidence_outputs/redone_MR_fulloutput.tsv")
res_subtypes <- read_tsv("../results/mr_evidence_outputs/mr_subtypes/all_traits_MR_vs_BCAC2020.tsv") %>% filter(outcome != "TNBC_BRCA1 ER-PR-HER-")
## category update 
res_subtypes <- res_subtypes %>% 
  rename(exposure.trait=exposure, exposure.id = id.exposure ) %>%
  create_exposure_categories() %>%  filter(exposure_cat != 'other') %>% 
  select(-exposure ) %>% 
  rename(id.exposure = exposure.id , exposure = exposure.trait) 


redone_MR %>% select(id.exposure, exposure_cat) %>% distinct() %>%  count(exposure_cat, sort=T)
res_subtypes  %>% select(id.exposure, exposure_cat) %>% distinct() %>% count(exposure_cat, sort=T)

all_mr_validation <- bind_rows(redone_MR, res_subtypes) %>% 
    filter(method %in% c('Inverse variance weighted', 'Wald ratio') ) %>% 
   arrange(pval) %>% mutate(qval = p.adjust(pval, method = "BH")) 

write_tsv(all_mr_validation, "../results/mr_evidence_outputs/all_multiple_testing.tsv")
  

tmp <- all_mr_validation %>% select(pval, qval) %>% 
  mutate(pval_sign = ifelse(pval <0.05, T, F)) %>% 
  mutate(qval_sign = ifelse(qval <0.05, T, F))

tmp %>% count(pval_sign) # FALSE	2406	 TRUE	728	
tmp %>% count(qval_sign) # FALSE	2802	TRUE	332	


x<- all_mr_validation %>%  filter(pval < 0.05) %>% select(id.exposure, exposure,outcome, exposure_cat, effect_direction) %>% distinct() 
length(unique(x$id.exposure)) # 213

y<- all_mr_validation %>%  filter(effect_direction != 'overlaps null') %>% select(id.exposure, exposure,outcome, exposure_cat, effect_direction) %>% distinct() 
length(unique(y$id.exposure)) # 213

z<- all_mr_validation %>%  filter(!is.na(id.outcome)) %>% filter(effect_direction != 'overlaps null') %>% select(id.exposure, exposure,outcome, exposure_cat, effect_direction) %>% distinct() 
length(unique(z$id.exposure)) # 171

w<- all_mr_validation %>%  filter(is.na(id.outcome)) %>% filter(effect_direction != 'overlaps null') %>% select(id.exposure, exposure,outcome, exposure_cat, effect_direction) %>% distinct() 
length(unique(w$id.exposure)) # 178

passed_pairs <- all_mr_validation %>%  mutate(mtc = ifelse(qval < 0.05, "*", NA))  %>%
  select(id.exposure, exposure,outcome, exposure_cat, effect_direction, OR_CI, pval, qval, mtc) %>% distinct() 

passed_pairs %>% filter(mtc == "*") %>% pull(id.exposure) %>% unique() %>% length() # 105
write_tsv(passed_pairs, "../results/mr_evidence_outputs/passed_multiple_testing.tsv")

```

### Sensitivity analysis review


# 1, >2 snps (so that sens tests are possible)
# 2. IVW must have effect 
# 3. egger and median must hve same effect direction
# 4. one egger or median also has to be signidicant

#. egger intecent must be low ~0.002 + large Pval =OK
#  flag low Q pval - is Q high? 



```{r}

mtc <- read_tsv("../results/mr_evidence_outputs/all_multiple_testing.tsv")  %>% 
  select(id.exposure, outcome,qval )%>% rename(pval_FDR_adj = qval)

## old outcomes
main_results <- read_tsv("../results/mr_evidence_outputs/redone_MR_fulloutput.tsv")
sens_results <- read_tsv("../results/mr_evidence_outputs/redone_MR_fulloutput_sens.tsv")

res_w_sens <- list()

outcomes1<- c('ieu-a-1126', 'ieu-a-1127', 'ieu-a-1128')

for (outcome_name in outcomes1){
  res_w_sens[[outcome_name]] <- add_sensitivity_cols(main_results, sens_results, outcome_name, mtc)
}

# new outcomes <- will add to the same list

main_results <- read_tsv("../results/mr_evidence_outputs/mr_subtypes/all_traits_MR_vs_BCAC2020.tsv") %>%  mutate(id.outcome = outcome)
sens_results <- read_tsv("../results/mr_evidence_outputs/mr_subtypes/all_traits_sensMR_vs_BCAC2020.tsv") %>%  mutate(id.outcome = outcome)

outcomes2 <- unique(main_results$id.outcome)

for (outcome_name in outcomes2){
  res_w_sens[[outcome_name]] <- add_sensitivity_cols(main_results, sens_results, outcome_name, mtc)
}

names(res_w_sens) <- c("All. BCAC 2017", "ER+. BCAC 2017 ", "ER-. BCAC 2017", 
                       "All. BCAC 2020", "Luminal A. BCAC 2020","Luminal B1. BCAC 2020", 
                                     "Luminal B2. BCAC 2020",  "HER2-enriched. BCAC 2020", "TNBC. BCAC 2020" )

res_w_sens <-res_w_sens[names(res_w_sens)[1:9]]# drop second TNBC

writexl::write_xlsx(res_w_sens, "../results/mr_evidence_outputs/all_data_with_sens_filters.xlsx")

## also save 
main_results17 <- read_tsv("../results/mr_evidence_outputs/redone_MR_fulloutput.tsv")
main_results20 <- read_tsv("../results/mr_evidence_outputs/mr_subtypes/all_traits_MR_vs_BCAC2020.tsv") %>%  mutate(id.outcome = outcome) %>% filter(outcome != "TNBC_BRCA1 ER-PR-HER-")

all_mr_tests <- bind_rows(main_results17,main_results20 ) %>% 
                arrange(exposure_cat, id.exposure) %>% 
                select(-exposure_name)

write_tsv(all_mr_tests, "../results/mr_evidence_outputs/all_data_validation.tsv")
```










# some testing



```{r}
### Subsetting to traits from MR that have enough lit space

lit_space_sizes <- read_tsv("../02_literature_related/literature_outputs/traits_marked_for_lit_analysis.tsv")

mr_results_df <- read_tsv("mr_evidence_outputs/trait_manual_ivw_subtypes_merged.tsv") # 314

emply_lit_spaces <- lit_space_sizes %>% filter(lit_analysis ==1 & lit_space_size == 0 ) 
excluded_from_lit_query  <- lit_space_sizes %>% filter(lit_analysis ==0 ) 


mr_results_sub <- mr_results_df %>% 
  filter(!id.exposure %in% c(emply_lit_spaces$id, excluded_from_lit_query$id)) %>% 
  tibble::column_to_rownames('id.exposure') %>%
  select(5:13) %>% 
  mutate(across(everything(), ~replace(., . ==  0 , NA))) %>% 
  filter(if_any(everything(), ~ !is.na(.))) %>% # drop those with NA in all outcomes
  mutate(across(everything(), ~replace_na(.x, 0))) %>% 
  tibble::rownames_to_column('id.exposure') %>% 
  left_join(mr_results_df %>% select(id.exposure, exposure, exposure_cat)) %>% 
  filter(exposure_cat != 'Lipids') %>% 
  select(exposure_cat, id.exposure, exposure, 
          "BCAC 2017"="ieu-a-1126",
          "BCAC 2020"="Breast cancer BCAC 2020",
          "ER+" = "ieu-a-1127" ,
          "ER-" = "ieu-a-1128" ,
          everything() )

# add protein names
protein_names <- read_csv("mr_evidence_outputs/protein_names.csv", col_names = F) %>% rename(name = X1, gene = X2)
mr_results_sub<- mr_results_sub %>% left_join(protein_names, by = c('exposure'='name')) 

# add lit size
mr_results_sub<- mr_results_sub %>% left_join(lit_space_sizes %>% select(id, unique_triples, unique_pairs), by =c("id.exposure"= 'id'))
  

# keep those that passed MTC
passed_pairs <- read_tsv("mr_evidence_outputs/passed_multiple_testing.tsv") %>% mutate(passed_MTC = T) %>% select(id.exposure, passed_MTC)
mr_results_sub<- mr_results_sub %>% left_join(passed_pairs) %>% # MB this is not a precise measure - MTC is different for different outcomes! 
                                        mutate_at(vars(passed_MTC), ~replace_na(.x, 0))
# add mediator counts 
med_counts <- read_tsv("mr_evidence_outputs/supl_table8_counts_med.csv") %>% select(exposure.id, med_count)
mr_results_sub<- mr_results_sub %>% left_join(med_counts, by=c("id.exposure"="exposure.id")) %>% distinct()


write_tsv(mr_results_sub, "mr_evidence_outputs/trait_manual_ivw_subtypes_merged_Litspaces_available_MTCpassed.tsv")


```

```{r}
# finding which proteins already appear in breast cancer lit space
common_terms_bc <- read_tsv("../02_literature_related/literature_outputs/breast_cancer_terms.tsv", col_names = T)

mr_results_sub %>%  filter(exposure_cat %in% c('Proteins', 'Metabolites')) %>% 
  mutate(in_bc_space = ifelse(gene %in% common_terms_bc$name | exposure %in% common_terms_bc$name, 1,0))%>%
  filter(in_bc_space ==1) %>% 
  left_join(common_terms_bc, by =c('exposure'='name')) %>% 
  left_join(common_terms_bc, by =c('gene'='name')) %>% 
  mutate(size_in_bc_space = coalesce(size.x, size.y)) %>% 
  select(-size.x, -size.y) %>% 
  View()
```




























