{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "38687\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import json\n",
    "from environs import Env\n",
    "from graph_functions import query_to_df, start_graph_session_crashdown_w_auth, is_it_cancer, is_it_breast_cancer\n",
    "from IPython.display import Image\n",
    "import requests\n",
    "\n",
    "env = Env()\n",
    "#session = start_graph_session(env) \n",
    "session = start_graph_session_crashdown_w_auth(env) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Extract genes expressed in breast tissue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "MATCH  (gene:Gene)-[gene_to_tissue:EXPRESSED_IN]->(tissue:Tissue) \n",
      "WHERE gene_to_tissue.tpm > 1\n",
      "AND  tissue.name = 'Breast - Mammary Tissue'\n",
      "RETURN gene.name, gene_to_tissue.tpm as TPM\n",
      "ORDER BY TPM desc\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene.name</th>\n",
       "      <th>TPM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MT-ND4</td>\n",
       "      <td>32170.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MT-CO2</td>\n",
       "      <td>28380.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FTL</td>\n",
       "      <td>5355.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>FABP4</td>\n",
       "      <td>4585.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RPS18</td>\n",
       "      <td>3804.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16907</th>\n",
       "      <td>RP11-800A3.4</td>\n",
       "      <td>1.00245</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16908</th>\n",
       "      <td>OAZ3</td>\n",
       "      <td>1.00150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16909</th>\n",
       "      <td>RP11-539I5.1</td>\n",
       "      <td>1.00150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16910</th>\n",
       "      <td>RP11-757O6.1</td>\n",
       "      <td>1.00110</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16911</th>\n",
       "      <td>RP1-261D10.2</td>\n",
       "      <td>1.00055</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16912 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          gene.name          TPM\n",
       "0            MT-ND4  32170.00000\n",
       "1            MT-CO2  28380.00000\n",
       "2               FTL   5355.00000\n",
       "3             FABP4   4585.00000\n",
       "4             RPS18   3804.00000\n",
       "...             ...          ...\n",
       "16907  RP11-800A3.4      1.00245\n",
       "16908          OAZ3      1.00150\n",
       "16909  RP11-539I5.1      1.00150\n",
       "16910  RP11-757O6.1      1.00110\n",
       "16911  RP1-261D10.2      1.00055\n",
       "\n",
       "[16912 rows x 2 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "query = \"\"\"\n",
    "MATCH  (gene:Gene)-[gene_to_tissue:EXPRESSED_IN]->(tissue:Tissue) \n",
    "WHERE gene_to_tissue.tpm > 1\n",
    "AND  tissue.name = 'Breast - Mammary Tissue'\n",
    "RETURN gene.name, gene_to_tissue.tpm as TPM\n",
    "ORDER BY TPM desc\n",
    "\"\"\"\n",
    "\n",
    "df = query_to_df(session, query)\n",
    "\n",
    "df\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/ny19205/opt/anaconda3/envs/neo4j_build/lib/python3.7/site-packages/ipykernel_launcher.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  \n",
      "/Users/ny19205/opt/anaconda3/envs/neo4j_build/lib/python3.7/site-packages/ipykernel_launcher.py:8: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  \n",
      "/Users/ny19205/opt/anaconda3/envs/neo4j_build/lib/python3.7/site-packages/ipykernel_launcher.py:10: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  # Remove the CWD from sys.path while we load stuff.\n"
     ]
    },
    {
     "ename": "AttributeError",
     "evalue": "'Series' object has no attribute 'values_count'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-25-534142ec8d7a>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     10\u001b[0m         \u001b[0mdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'tpm_category'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mind\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"low\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     11\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 12\u001b[0;31m \u001b[0mdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'tpm_category'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvalues_count\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/opt/anaconda3/envs/neo4j_build/lib/python3.7/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m__getattr__\u001b[0;34m(self, name)\u001b[0m\n\u001b[1;32m   5463\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_info_axis\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_can_hold_identifiers_and_holds_name\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5464\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 5465\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mobject\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__getattribute__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   5466\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5467\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__setattr__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'Series' object has no attribute 'values_count'"
     ]
    }
   ],
   "source": [
    "# add expression level categories\n",
    "df['tpm_category'] = ''\n",
    "\n",
    "for ind in df.index:\n",
    "    if df['TPM'][ind] >= 100:\n",
    "        df['tpm_category'][ind] = \"high\"\n",
    "    elif df['TPM'][ind] < 100 and df['TPM'][ind] > 10 :\n",
    "        df['tpm_category'][ind] = \"medium\"\n",
    "    elif df['TPM'][ind] <= 10:\n",
    "        df['tpm_category'][ind] = \"low\"\n",
    "        \n",
    "df['tpm_category'].values_count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "medium    8039\n",
       "low       7918\n",
       "high       955\n",
       "Name: tpm_category, dtype: int64"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['tpm_category'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene.name</th>\n",
       "      <th>drug.label</th>\n",
       "      <th>d.label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NDUFB10</td>\n",
       "      <td>METFORMIN</td>\n",
       "      <td>acanthosis nigricans (disease)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NDUFB10</td>\n",
       "      <td>METFORMIN</td>\n",
       "      <td>altitude sickness</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EPHX1</td>\n",
       "      <td>CARBAMAZEPINE</td>\n",
       "      <td>trigeminal neuralgia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NDUFS5</td>\n",
       "      <td>METFORMIN</td>\n",
       "      <td>acanthosis nigricans (disease)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NDUFS5</td>\n",
       "      <td>METFORMIN</td>\n",
       "      <td>altitude sickness</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>613</th>\n",
       "      <td>CDK7</td>\n",
       "      <td>ZOTIRACICLIB</td>\n",
       "      <td>gliosarcoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>614</th>\n",
       "      <td>CDK7</td>\n",
       "      <td>ZOTIRACICLIB</td>\n",
       "      <td>rectal cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>615</th>\n",
       "      <td>CDK7</td>\n",
       "      <td>SELICICLIB</td>\n",
       "      <td>ACTH-dependent Cushing syndrome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>616</th>\n",
       "      <td>CDK7</td>\n",
       "      <td>AT-7519</td>\n",
       "      <td>mantle cell lymphoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>617</th>\n",
       "      <td>CDK7</td>\n",
       "      <td>ALVOCIDIB</td>\n",
       "      <td>endometrial cancer</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>618 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    gene.name     drug.label                          d.label\n",
       "0     NDUFB10      METFORMIN   acanthosis nigricans (disease)\n",
       "1     NDUFB10      METFORMIN                altitude sickness\n",
       "2       EPHX1  CARBAMAZEPINE             trigeminal neuralgia\n",
       "3      NDUFS5      METFORMIN   acanthosis nigricans (disease)\n",
       "4      NDUFS5      METFORMIN                altitude sickness\n",
       "..        ...            ...                              ...\n",
       "613      CDK7   ZOTIRACICLIB                      gliosarcoma\n",
       "614      CDK7   ZOTIRACICLIB                    rectal cancer\n",
       "615      CDK7     SELICICLIB  ACTH-dependent Cushing syndrome\n",
       "616      CDK7        AT-7519             mantle cell lymphoma\n",
       "617      CDK7      ALVOCIDIB               endometrial cancer\n",
       "\n",
       "[618 rows x 3 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get drugs and diseases\n",
    "gene_list1 = list(set(df['gene.name']))\n",
    "\n",
    "query=\"\"\"\n",
    "match (gene:Gene)-[]-(drug:Drug)-[]-(d:Disease)\n",
    "where gene.name in {}\n",
    "return gene.name , drug.label, d.label\n",
    "\"\"\".format(gene_list1)\n",
    "\n",
    "exp_to_disease = query_to_df(session, query, False)\n",
    "exp_to_disease"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene.name</th>\n",
       "      <th>d.label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SSTR1</td>\n",
       "      <td>neuroendocrine neoplasm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SSTR1</td>\n",
       "      <td>ACTH-dependent Cushing syndrome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SSTR1</td>\n",
       "      <td>acromegaly</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FOLR1</td>\n",
       "      <td>endometrial clear cell adenocarcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>FOLR1</td>\n",
       "      <td>endometrial serous adenocarcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>613</th>\n",
       "      <td>ADRA2A</td>\n",
       "      <td>ocular hypertension</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>614</th>\n",
       "      <td>FN1</td>\n",
       "      <td>soft tissue sarcoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>615</th>\n",
       "      <td>ITGAL</td>\n",
       "      <td>dry eye syndrome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>616</th>\n",
       "      <td>NDUFS8</td>\n",
       "      <td>acanthosis nigricans (disease)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>617</th>\n",
       "      <td>NDUFS8</td>\n",
       "      <td>altitude sickness</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>554 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    gene.name                                d.label\n",
       "0       SSTR1                neuroendocrine neoplasm\n",
       "1       SSTR1        ACTH-dependent Cushing syndrome\n",
       "2       SSTR1                             acromegaly\n",
       "4       FOLR1  endometrial clear cell adenocarcinoma\n",
       "5       FOLR1      endometrial serous adenocarcinoma\n",
       "..        ...                                    ...\n",
       "613    ADRA2A                    ocular hypertension\n",
       "614       FN1                    soft tissue sarcoma\n",
       "615     ITGAL                       dry eye syndrome\n",
       "616    NDUFS8         acanthosis nigricans (disease)\n",
       "617    NDUFS8                      altitude sickness\n",
       "\n",
       "[554 rows x 2 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# genes and diseases they target\n",
    "exp_to_disease2 = exp_to_disease[['gene.name', 'd.label']].drop_duplicates()\n",
    "exp_to_disease2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "# is it cancer\n",
    "exp_to_disease2 = is_it_cancer(exp_to_disease2,'d.label' )\n",
    "exp_to_disease2 = is_it_breast_cancer(exp_to_disease2,'d.label' )\n",
    "exp_to_disease2\n",
    "exp_to_disease2.to_csv(\"../query_results/test.tsv\", sep=\"\\t\", index=True)\n",
    "\n",
    "#out = pd.DataFrame(pd.Series(exp_to_disease2.groupby([\"gene.name\", \"is_it_cancer\"]).size()))\n",
    "#out\n",
    "#out.to_csv(\"../query_results/test.tsv\", sep=\"\\t\", index=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### GWAS to genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "MATCH (gwas:Gwas) -[gwas_to_variant:GWAS_TO_VARIANT]->(variant:Variant)-[variant_to_gene:VARIANT_TO_GENE]-> (gene:Gene)\n",
      "WHERE gwas.id = 'ieu-a-1126' \n",
      "AND gene.name is not null \n",
      "RETURN  variant.name, gene.name\n",
      "\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant.name</th>\n",
       "      <th>gene.name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs532436</td>\n",
       "      <td>ABO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs28512361</td>\n",
       "      <td>WI2-85898F10.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs5995875</td>\n",
       "      <td>MKL1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs5995875</td>\n",
       "      <td>COX6B1P3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs4820318</td>\n",
       "      <td>PLA2G6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>rs1230666</td>\n",
       "      <td>MAGI3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>rs1707302</td>\n",
       "      <td>PIK3R3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>rs1707302</td>\n",
       "      <td>RP4-533D7.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>rs2992756</td>\n",
       "      <td>KLHDC7A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>rs2506889</td>\n",
       "      <td>PEX14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>137 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    variant.name       gene.name\n",
       "0       rs532436             ABO\n",
       "1     rs28512361  WI2-85898F10.1\n",
       "2      rs5995875            MKL1\n",
       "3      rs5995875        COX6B1P3\n",
       "4      rs4820318          PLA2G6\n",
       "..           ...             ...\n",
       "132    rs1230666           MAGI3\n",
       "133    rs1707302          PIK3R3\n",
       "134    rs1707302     RP4-533D7.5\n",
       "135    rs2992756         KLHDC7A\n",
       "136    rs2506889           PEX14\n",
       "\n",
       "[137 rows x 2 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"\"\"\n",
    "MATCH (gwas:Gwas) -[gwas_to_variant:GWAS_TO_VARIANT]->(variant:Variant)-[variant_to_gene:VARIANT_TO_GENE]-> (gene:Gene)\n",
    "WHERE gwas.id = '{gwas_id}' \n",
    "AND gene.name is not null \n",
    "RETURN  variant.name, gene.name\n",
    "\n",
    "\"\"\".format(gwas_id='ieu-a-1126')\n",
    "\n",
    "df = query_to_df(session, query)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "match (gene:Gene)-[]-(drug:Drug)-[]-(d:Disease)\n",
      "where gene.name in ['PLA2G6', 'AC092594.1', 'TTC28', 'PRC1', 'KRTCAP2', 'ITPR1', 'KLHDC7A', 'AQP4-AS1', 'HSCB', 'COX6B1P3', 'CHEK2', 'CUX1', 'HSPA4', 'SETBP1', 'MLLT10', 'VGLL3', 'Y_RNA', 'BRCA2', 'RPL30P5', 'TGFBR2', 'TET2', 'GRHL1', 'GATAD2A', 'MIR642A', 'ZMIZ1', 'GAREM', 'RAD51B', 'RP4-533D7.5', 'OTUD7B', 'PIDD', 'AMFR', 'OR2A1-AS1', 'GIPR', 'ATG10', 'RP11-201K10.3', 'ADSSL1', 'NREP', 'RP11-408H1.3', 'RNU1-58P', 'RCCD1', 'ASTN2', 'RP11-266N13.2', 'ARHGEF5', 'AC090453.1', 'RP11-127H5.1', 'KCNN4', 'LMX1B', 'FAM175A', 'MIER3', 'MCM8', 'PIK3R3', 'TRIM46', 'TLR1', 'CMSS1', 'TCF7L2', 'SLC25A21', 'ESR1', 'CASC8', 'PHLDA3', 'ATXN7', 'DNAJC1', 'SLC14A2', 'NEK10', 'CDCA7L', 'L3MBTL3', 'MAGI3', 'CHST9', 'ZNF365', 'RP11-134G8.8', 'RIN3', 'PDE4D', 'SLC6A18', 'AC104801.1', 'MTMR11', 'DIRC3', 'KANSL1', 'CBX8', 'RP11-438P9.2', 'ZFPM2', 'TERT', 'AC016644.1', 'SSBP4', 'EXO1', 'PRC1-AS1', 'CASC9', 'AC010969.1', 'AC012363.13', 'RANBP9', 'PEX14', 'MKL1', 'ZBTB38', 'CASP8', 'AP000439.5', 'EBF1', 'ATAD5', 'WI2-85898F10.1', 'STXBP4', 'FGFR2', 'FAM46A', 'EMBP1', 'CASC16', 'CDYL2', 'MIR1208', 'CREB5', 'PAX9', 'DNAH11', 'CD160', 'MIR640', 'LINC00536', 'RP11-145E5.5', 'RP11-353N14.1', 'CASC19', 'CCDC88C', 'FTO', 'ESRRG', 'ADCY3', 'RP11-10L7.1', 'RP5-967N21.7', 'RP11-382A18.2', 'N4BP2L1', 'LINC-PINT', 'ABO', 'MYEOV', 'SLC25A22', 'RNF115', 'SUB1', 'FILIP1L', 'LSP1', 'ADAM29']\n",
      "return gene.name , drug.label, d.label\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene.name</th>\n",
       "      <th>drug.label</th>\n",
       "      <th>d.label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>KCNN4</td>\n",
       "      <td>CHLORZOXAZONE</td>\n",
       "      <td>osteoarthritis, hip</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PIK3R3</td>\n",
       "      <td>DS-7423</td>\n",
       "      <td>endometrial cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PIK3R3</td>\n",
       "      <td>SAMOTOLISIB</td>\n",
       "      <td>endometrial cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PIK3R3</td>\n",
       "      <td>PILARALISIB</td>\n",
       "      <td>endometrial cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PDE4D</td>\n",
       "      <td>PENTOXIFYLLINE</td>\n",
       "      <td>lytic metastatic bone lesion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>PDE4D</td>\n",
       "      <td>APREMILAST</td>\n",
       "      <td>parapsoriasis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CASP8</td>\n",
       "      <td>EMRICASAN</td>\n",
       "      <td>non-alcoholic steatohepatitis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CASP8</td>\n",
       "      <td>NIVOCASAN</td>\n",
       "      <td>non-alcoholic steatohepatitis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>FGFR2</td>\n",
       "      <td>DERAZANTINIB</td>\n",
       "      <td>intrahepatic cholangiocarcinoma</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene.name      drug.label                          d.label\n",
       "0     KCNN4   CHLORZOXAZONE              osteoarthritis, hip\n",
       "1    PIK3R3         DS-7423               endometrial cancer\n",
       "2    PIK3R3     SAMOTOLISIB               endometrial cancer\n",
       "3    PIK3R3     PILARALISIB               endometrial cancer\n",
       "4     PDE4D  PENTOXIFYLLINE     lytic metastatic bone lesion\n",
       "5     PDE4D      APREMILAST                    parapsoriasis\n",
       "6     CASP8       EMRICASAN    non-alcoholic steatohepatitis\n",
       "7     CASP8       NIVOCASAN    non-alcoholic steatohepatitis\n",
       "8     FGFR2    DERAZANTINIB  intrahepatic cholangiocarcinoma"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get drugs and diseases\n",
    "gene_list2 = list(set(df['gene.name']))\n",
    "\n",
    "query=\"\"\"\n",
    "match (gene:Gene)-[]-(drug:Drug)-[]-(d:Disease)\n",
    "where gene.name in {}\n",
    "return gene.name , drug.label, d.label\n",
    "\"\"\".format(gene_list2)\n",
    "\n",
    "exp_to_disease = query_to_df(session, query, True)\n",
    "exp_to_disease"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene.name</th>\n",
       "      <th>d.label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>KCNN4</td>\n",
       "      <td>osteoarthritis, hip</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PDE4D</td>\n",
       "      <td>lytic metastatic bone lesion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PDE4D</td>\n",
       "      <td>parapsoriasis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CASP8</td>\n",
       "      <td>non-alcoholic steatohepatitis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>PIK3R3</td>\n",
       "      <td>endometrial cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>FGFR2</td>\n",
       "      <td>intrahepatic cholangiocarcinoma</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene.name                          d.label\n",
       "0     KCNN4              osteoarthritis, hip\n",
       "1     PDE4D     lytic metastatic bone lesion\n",
       "2     PDE4D                    parapsoriasis\n",
       "3     CASP8    non-alcoholic steatohepatitis\n",
       "5    PIK3R3               endometrial cancer\n",
       "8     FGFR2  intrahepatic cholangiocarcinoma"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# genes and diseases they target\n",
    "exp_to_disease2 = exp_to_disease[['gene.name', 'd.label']].drop_duplicates()\n",
    "exp_to_disease2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### test clinvar connections"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene.name</th>\n",
       "      <th>d.label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DHX38</td>\n",
       "      <td>retinitis pigmentosa 84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MASP2</td>\n",
       "      <td>immunodeficiency due to MASP-2 deficiency</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TP63</td>\n",
       "      <td>orofacial cleft</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TP63</td>\n",
       "      <td>orofacial cleft 8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TP63</td>\n",
       "      <td>limb-mammary syndrome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5669</th>\n",
       "      <td>WIPI2</td>\n",
       "      <td>intellectual developmental disorder with short...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5670</th>\n",
       "      <td>FH</td>\n",
       "      <td>hereditary leiomyomatosis and renal cell cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5671</th>\n",
       "      <td>NDUFS8</td>\n",
       "      <td>mitochondrial complex 1 deficiency, nuclear ty...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5672</th>\n",
       "      <td>TBC1D23</td>\n",
       "      <td>pontocerebellar hypoplasia, type 11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5673</th>\n",
       "      <td>ARHGEF2</td>\n",
       "      <td>neurodevelopmental disorder with midbrain and ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5674 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene.name                                            d.label\n",
       "0        DHX38                            retinitis pigmentosa 84\n",
       "1        MASP2          immunodeficiency due to MASP-2 deficiency\n",
       "2         TP63                                    orofacial cleft\n",
       "3         TP63                                  orofacial cleft 8\n",
       "4         TP63                              limb-mammary syndrome\n",
       "...        ...                                                ...\n",
       "5669     WIPI2  intellectual developmental disorder with short...\n",
       "5670        FH    hereditary leiomyomatosis and renal cell cancer\n",
       "5671    NDUFS8  mitochondrial complex 1 deficiency, nuclear ty...\n",
       "5672   TBC1D23                pontocerebellar hypoplasia, type 11\n",
       "5673   ARHGEF2  neurodevelopmental disorder with midbrain and ...\n",
       "\n",
       "[5674 rows x 2 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query=\"\"\"\n",
    "match (gene:Gene)-[]-(d:Disease)\n",
    "where gene.name in {}\n",
    "return gene.name , d.label\n",
    "\"\"\".format(gene_list1) # exp genes\n",
    "\n",
    "y= query_to_df(session, query, False)\n",
    "y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "y = is_it_cancer(y,'d.label' )\n",
    "y = is_it_breast_cancer(y,'d.label' )\n",
    "y[y['is_it_cancer']=='yes']\n",
    "y.to_csv(\"../query_results/test.tsv\", sep=\"\\t\", index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene.name</th>\n",
       "      <th>d.label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PLA2G6</td>\n",
       "      <td>PLA2G6-associated neurodegeneration</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PLA2G6</td>\n",
       "      <td>neurodegeneration with brain iron accumulation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PLA2G6</td>\n",
       "      <td>autosomal recessive Parkinson disease 14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PLA2G6</td>\n",
       "      <td>Parkinson disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ITPR1</td>\n",
       "      <td>aniridia-cerebellar ataxia-intellectual disabi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>CCDC88C</td>\n",
       "      <td>Hydrocephalus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>CCDC88C</td>\n",
       "      <td>congenital hydrocephalus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>CCDC88C</td>\n",
       "      <td>hydrocephalus, nonsyndromic, autosomal recessi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>CCDC88C</td>\n",
       "      <td>spinocerebellar ataxia type 40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>SLC25A22</td>\n",
       "      <td>early myoclonic encephalopathy</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>93 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene.name                                            d.label\n",
       "0     PLA2G6                PLA2G6-associated neurodegeneration\n",
       "1     PLA2G6     neurodegeneration with brain iron accumulation\n",
       "2     PLA2G6           autosomal recessive Parkinson disease 14\n",
       "3     PLA2G6                                  Parkinson disease\n",
       "4      ITPR1  aniridia-cerebellar ataxia-intellectual disabi...\n",
       "..       ...                                                ...\n",
       "88   CCDC88C                                      Hydrocephalus\n",
       "89   CCDC88C                           congenital hydrocephalus\n",
       "90   CCDC88C  hydrocephalus, nonsyndromic, autosomal recessi...\n",
       "91   CCDC88C                     spinocerebellar ataxia type 40\n",
       "92  SLC25A22                     early myoclonic encephalopathy\n",
       "\n",
       "[93 rows x 2 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#gwas vars\n",
    "query=\"\"\"\n",
    "match (gene:Gene)-[]-(d:Disease)\n",
    "where gene.name in {}\n",
    "return gene.name , d.label\n",
    "\"\"\".format(gene_list2)# GWAS gens\n",
    "\n",
    "x= query_to_df(session, query, False)\n",
    "x"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene.name</th>\n",
       "      <th>d.label</th>\n",
       "      <th>is_it_cancer</th>\n",
       "      <th>is_it_breast_cancer</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CHEK2</td>\n",
       "      <td>breast neoplasm</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>CHEK2</td>\n",
       "      <td>prostate cancer</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>CHEK2</td>\n",
       "      <td>hereditary breast carcinoma</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>MLLT10</td>\n",
       "      <td>acute myeloid leukemia</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>BRCA2</td>\n",
       "      <td>breast neoplasm</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>BRCA2</td>\n",
       "      <td>kidney Wilms tumor</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>BRCA2</td>\n",
       "      <td>Wilms tumor 1</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>BRCA2</td>\n",
       "      <td>hereditary breast ovarian cancer syndrome</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>BRCA2</td>\n",
       "      <td>prostate cancer</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>BRCA2</td>\n",
       "      <td>hereditary breast carcinoma</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>TGFBR2</td>\n",
       "      <td>familial colorectal cancer</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>TGFBR2</td>\n",
       "      <td>colorectal cancer, hereditary nonpolyposis, ty...</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>TGFBR2</td>\n",
       "      <td>esophageal cancer</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>TGFBR2</td>\n",
       "      <td>colon carcinoma</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>ZMIZ1</td>\n",
       "      <td>neurodevelopmental disorder with dysmorphic fa...</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>ESR1</td>\n",
       "      <td>breast neoplasm</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>ESR1</td>\n",
       "      <td>hereditary breast carcinoma</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>TERT</td>\n",
       "      <td>acute myeloid leukemia</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>TERT</td>\n",
       "      <td>melanoma (disease)</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>TERT</td>\n",
       "      <td>cutaneous melanoma (disease)</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>CASP8</td>\n",
       "      <td>breast neoplasm</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>CASP8</td>\n",
       "      <td>lung cancer</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>CASP8</td>\n",
       "      <td>hereditary breast carcinoma</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>FGFR2</td>\n",
       "      <td>Antley-Bixler syndrome without genital anomali...</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene.name                                            d.label is_it_cancer  \\\n",
       "7      CHEK2                                    breast neoplasm          yes   \n",
       "10     CHEK2                                    prostate cancer          yes   \n",
       "11     CHEK2                        hereditary breast carcinoma          yes   \n",
       "15    MLLT10                             acute myeloid leukemia          yes   \n",
       "16     BRCA2                                    breast neoplasm          yes   \n",
       "19     BRCA2                                 kidney Wilms tumor          yes   \n",
       "22     BRCA2                                      Wilms tumor 1          yes   \n",
       "24     BRCA2          hereditary breast ovarian cancer syndrome          yes   \n",
       "25     BRCA2                                    prostate cancer          yes   \n",
       "26     BRCA2                        hereditary breast carcinoma          yes   \n",
       "28    TGFBR2                         familial colorectal cancer          yes   \n",
       "31    TGFBR2  colorectal cancer, hereditary nonpolyposis, ty...          yes   \n",
       "33    TGFBR2                                  esophageal cancer          yes   \n",
       "34    TGFBR2                                    colon carcinoma          yes   \n",
       "37     ZMIZ1  neurodevelopmental disorder with dysmorphic fa...          yes   \n",
       "44      ESR1                                    breast neoplasm          yes   \n",
       "47      ESR1                        hereditary breast carcinoma          yes   \n",
       "62      TERT                             acute myeloid leukemia          yes   \n",
       "63      TERT                                 melanoma (disease)          yes   \n",
       "64      TERT                       cutaneous melanoma (disease)          yes   \n",
       "66     CASP8                                    breast neoplasm          yes   \n",
       "67     CASP8                                        lung cancer          yes   \n",
       "71     CASP8                        hereditary breast carcinoma          yes   \n",
       "74     FGFR2  Antley-Bixler syndrome without genital anomali...          yes   \n",
       "\n",
       "   is_it_breast_cancer  \n",
       "7                  yes  \n",
       "10                  no  \n",
       "11                 yes  \n",
       "15                  no  \n",
       "16                 yes  \n",
       "19                  no  \n",
       "22                  no  \n",
       "24                 yes  \n",
       "25                  no  \n",
       "26                 yes  \n",
       "28                  no  \n",
       "31                  no  \n",
       "33                  no  \n",
       "34                  no  \n",
       "37                  no  \n",
       "44                 yes  \n",
       "47                 yes  \n",
       "62                  no  \n",
       "63                  no  \n",
       "64                  no  \n",
       "66                 yes  \n",
       "67                  no  \n",
       "71                 yes  \n",
       "74                  no  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x = is_it_cancer(x,'d.label' )\n",
    "x = is_it_breast_cancer(x,'d.label' )\n",
    "x[x['is_it_cancer']=='yes']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Which of breast cancer clivar genes also occur in GWAS/expr data? "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>d.label</th>\n",
       "      <th>d.id</th>\n",
       "      <th>is_it_cancer</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>malignant breast melanoma</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MONDO_0002975</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>breast rhabdomyosarcoma</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MONDO_0002859</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>breast mucinous carcinoma</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MONDO_0002707</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>breast mucinous cystadenocarcinoma</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MONDO_0002705</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>signet ring cell breast carcinoma</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MONDO_0002671</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>165</th>\n",
       "      <td>breast-ovarian cancer, familial, susceptibilit...</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MONDO_0011450</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167</th>\n",
       "      <td>triple-negative breast carcinoma</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MONDO_0005494</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>female breast central part cancer</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MONDO_0001805</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169</th>\n",
       "      <td>female breast lower-outer quadrant cancer</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MONDO_0001850</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>solid papillary breast carcinoma</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MONDO_0044783</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>105 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               d.label  \\\n",
       "0                            malignant breast melanoma   \n",
       "1                              breast rhabdomyosarcoma   \n",
       "2                            breast mucinous carcinoma   \n",
       "3                   breast mucinous cystadenocarcinoma   \n",
       "5                    signet ring cell breast carcinoma   \n",
       "..                                                 ...   \n",
       "165  breast-ovarian cancer, familial, susceptibilit...   \n",
       "167                   triple-negative breast carcinoma   \n",
       "168                  female breast central part cancer   \n",
       "169          female breast lower-outer quadrant cancer   \n",
       "170                   solid papillary breast carcinoma   \n",
       "\n",
       "                                             d.id is_it_cancer  \n",
       "0    http://purl.obolibrary.org/obo/MONDO_0002975          yes  \n",
       "1    http://purl.obolibrary.org/obo/MONDO_0002859          yes  \n",
       "2    http://purl.obolibrary.org/obo/MONDO_0002707          yes  \n",
       "3    http://purl.obolibrary.org/obo/MONDO_0002705          yes  \n",
       "5    http://purl.obolibrary.org/obo/MONDO_0002671          yes  \n",
       "..                                            ...          ...  \n",
       "165  http://purl.obolibrary.org/obo/MONDO_0011450          yes  \n",
       "167  http://purl.obolibrary.org/obo/MONDO_0005494          yes  \n",
       "168  http://purl.obolibrary.org/obo/MONDO_0001805          yes  \n",
       "169  http://purl.obolibrary.org/obo/MONDO_0001850          yes  \n",
       "170  http://purl.obolibrary.org/obo/MONDO_0044783          yes  \n",
       "\n",
       "[105 rows x 3 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query=\"\"\"\n",
    "match (d:Disease)\n",
    "where toLower(d.label) contains \"breast\"\n",
    "return  d.label, d.id\n",
    "\"\"\"\n",
    "x= query_to_df(session, query, False)\n",
    "y = is_it_cancer(x,'d.label' )\n",
    "y['is_it_cancer'].value_counts()\n",
    "#yes    114\n",
    "#no      57\n",
    "\n",
    "y_cancer = y[y['is_it_cancer']=='yes']\n",
    "y_cancer\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "known clinical var genes\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['CASP8',\n",
       " 'BRCA1',\n",
       " 'CHEK2',\n",
       " 'PPM1D',\n",
       " 'RAD54L',\n",
       " 'RAD51C',\n",
       " 'NQO2',\n",
       " 'ATM',\n",
       " 'BARD1',\n",
       " 'SLC22A18',\n",
       " 'RAD51',\n",
       " 'PTEN',\n",
       " 'XRCC3',\n",
       " 'AKT1',\n",
       " 'RB1CC1',\n",
       " 'BRCA2',\n",
       " 'TP53',\n",
       " 'BRIP1',\n",
       " 'KRAS',\n",
       " 'HMMR',\n",
       " 'CDH1',\n",
       " 'PHB',\n",
       " 'RAD51D',\n",
       " 'STK11',\n",
       " 'ESR1',\n",
       " 'PALB2',\n",
       " 'PIK3CA']"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# map diseases to genes\n",
    "bc_ids = list(set(y_cancer['d.id']))\n",
    "query=\"\"\"\n",
    "match (gene:Gene)-[]-(d:Disease)\n",
    "where d.id in {}\n",
    "return  d.label, d.id,gene.name \n",
    "\"\"\".format(bc_ids) \n",
    "x= query_to_df(session, query, False)\n",
    "x\n",
    "\n",
    "print(\"known clinical var genes\")\n",
    "clinvar_bc_genes = list(set(x['gene.name']))\n",
    "clinvar_bc_genes##27"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene.name</th>\n",
       "      <th>drugs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CASP8</td>\n",
       "      <td>[EMRICASAN, NIVOCASAN]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHEK2</td>\n",
       "      <td>[XL-844, PREXASERTIB]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AKT1</td>\n",
       "      <td>[MSC-2363318A, XL-418, BAY-1125976, GSK-690693...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ESR1</td>\n",
       "      <td>[ACOLBIFENE, GTX-758, ESTRONE, SR16234, ESTROG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PIK3CA</td>\n",
       "      <td>[WX-037, PA-799, BGT-226, BAY-1082439, PKI-179...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene.name                                              drugs\n",
       "0     CASP8                             [EMRICASAN, NIVOCASAN]\n",
       "1     CHEK2                              [XL-844, PREXASERTIB]\n",
       "2      AKT1  [MSC-2363318A, XL-418, BAY-1125976, GSK-690693...\n",
       "3      ESR1  [ACOLBIFENE, GTX-758, ESTRONE, SR16234, ESTROG...\n",
       "4    PIK3CA  [WX-037, PA-799, BGT-226, BAY-1082439, PKI-179..."
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check which of these genes are drug targets\n",
    "\n",
    "query=\"\"\"\n",
    "match (gene:Gene)-[]-(drug:Drug)\n",
    "where gene.name in {}\n",
    "return  gene.name , collect(drug.label) as drugs\n",
    "\"\"\".format(clinvar_bc_genes) \n",
    "x= query_to_df(session, query, False)\n",
    "x #- of those, 5 are drug targets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>drugs</th>\n",
       "      <th>d.label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EMRICASAN</td>\n",
       "      <td>non-alcoholic steatohepatitis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NIVOCASAN</td>\n",
       "      <td>non-alcoholic steatohepatitis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AFURESERTIB</td>\n",
       "      <td>Langerhans cell histiocytosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DS-7423</td>\n",
       "      <td>endometrial cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PILARALISIB</td>\n",
       "      <td>endometrial cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SAMOTOLISIB</td>\n",
       "      <td>endometrial cancer</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         drugs                        d.label\n",
       "0    EMRICASAN  non-alcoholic steatohepatitis\n",
       "1    NIVOCASAN  non-alcoholic steatohepatitis\n",
       "2  AFURESERTIB  Langerhans cell histiocytosis\n",
       "3      DS-7423             endometrial cancer\n",
       "4  PILARALISIB             endometrial cancer\n",
       "5  SAMOTOLISIB             endometrial cancer"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# repeat bove just extarcting drugs ----- missing lots of link to BC\n",
    "query=\"\"\"\n",
    "match (gene:Gene)-[]-(drug:Drug)-[]-(d:Disease)\n",
    "where gene.name in {}\n",
    "return distinct  drug.label as drugs, d.label\n",
    "\"\"\".format(clinvar_bc_genes) \n",
    "x= query_to_df(session, query, False)\n",
    "x # there are 70 drugs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene.name</th>\n",
       "      <th>TPM</th>\n",
       "      <th>tpm_category</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>744</th>\n",
       "      <td>PHB</td>\n",
       "      <td>124.4500</td>\n",
       "      <td>high</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1870</th>\n",
       "      <td>AKT1</td>\n",
       "      <td>59.0550</td>\n",
       "      <td>medium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2673</th>\n",
       "      <td>STK11</td>\n",
       "      <td>44.1250</td>\n",
       "      <td>medium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3189</th>\n",
       "      <td>PTEN</td>\n",
       "      <td>38.2800</td>\n",
       "      <td>medium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3632</th>\n",
       "      <td>TP53</td>\n",
       "      <td>33.8150</td>\n",
       "      <td>medium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3914</th>\n",
       "      <td>NQO2</td>\n",
       "      <td>31.4550</td>\n",
       "      <td>medium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4081</th>\n",
       "      <td>CDH1</td>\n",
       "      <td>30.1350</td>\n",
       "      <td>medium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5524</th>\n",
       "      <td>RB1CC1</td>\n",
       "      <td>21.5400</td>\n",
       "      <td>medium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6953</th>\n",
       "      <td>PPM1D</td>\n",
       "      <td>15.9150</td>\n",
       "      <td>medium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7223</th>\n",
       "      <td>PIK3CA</td>\n",
       "      <td>14.9700</td>\n",
       "      <td>medium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7847</th>\n",
       "      <td>KRAS</td>\n",
       "      <td>13.0850</td>\n",
       "      <td>medium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8397</th>\n",
       "      <td>XRCC3</td>\n",
       "      <td>11.4850</td>\n",
       "      <td>medium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8939</th>\n",
       "      <td>SLC22A18</td>\n",
       "      <td>10.1400</td>\n",
       "      <td>medium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9485</th>\n",
       "      <td>ATM</td>\n",
       "      <td>8.9235</td>\n",
       "      <td>low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9679</th>\n",
       "      <td>RAD51C</td>\n",
       "      <td>8.4875</td>\n",
       "      <td>low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10382</th>\n",
       "      <td>CASP8</td>\n",
       "      <td>7.1575</td>\n",
       "      <td>low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10637</th>\n",
       "      <td>RAD51D</td>\n",
       "      <td>6.6420</td>\n",
       "      <td>low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11003</th>\n",
       "      <td>PALB2</td>\n",
       "      <td>6.0000</td>\n",
       "      <td>low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11171</th>\n",
       "      <td>ESR1</td>\n",
       "      <td>5.7375</td>\n",
       "      <td>low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12153</th>\n",
       "      <td>CHEK2</td>\n",
       "      <td>4.2160</td>\n",
       "      <td>low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12758</th>\n",
       "      <td>BARD1</td>\n",
       "      <td>3.5165</td>\n",
       "      <td>low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15580</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>1.4800</td>\n",
       "      <td>low</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene.name       TPM tpm_category\n",
       "744         PHB  124.4500         high\n",
       "1870       AKT1   59.0550       medium\n",
       "2673      STK11   44.1250       medium\n",
       "3189       PTEN   38.2800       medium\n",
       "3632       TP53   33.8150       medium\n",
       "3914       NQO2   31.4550       medium\n",
       "4081       CDH1   30.1350       medium\n",
       "5524     RB1CC1   21.5400       medium\n",
       "6953      PPM1D   15.9150       medium\n",
       "7223     PIK3CA   14.9700       medium\n",
       "7847       KRAS   13.0850       medium\n",
       "8397      XRCC3   11.4850       medium\n",
       "8939   SLC22A18   10.1400       medium\n",
       "9485        ATM    8.9235          low\n",
       "9679     RAD51C    8.4875          low\n",
       "10382     CASP8    7.1575          low\n",
       "10637    RAD51D    6.6420          low\n",
       "11003     PALB2    6.0000          low\n",
       "11171      ESR1    5.7375          low\n",
       "12153     CHEK2    4.2160          low\n",
       "12758     BARD1    3.5165          low\n",
       "15580     BRCA1    1.4800          low"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# compare with expression genes\n",
    "df # from first chunk\n",
    "\n",
    "df[df['gene.name'].isin(clinvar_bc_genes)] # 22 /27\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant.name</th>\n",
       "      <th>gene.name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>rs5997389</td>\n",
       "      <td>CHEK2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>rs11571833</td>\n",
       "      <td>BRCA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>rs2747652</td>\n",
       "      <td>ESR1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>rs76956704</td>\n",
       "      <td>ESR1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113</th>\n",
       "      <td>rs3769821</td>\n",
       "      <td>CASP8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    variant.name gene.name\n",
       "5      rs5997389     CHEK2\n",
       "43    rs11571833     BRCA2\n",
       "79     rs2747652      ESR1\n",
       "80    rs76956704      ESR1\n",
       "113    rs3769821     CASP8"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# compare with GWAS genes\n",
    "df # from second chunk\n",
    "\n",
    "df[df['gene.name'].isin(clinvar_bc_genes)] # 4 /27\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
