{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "38687\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import json\n",
    "from environs import Env\n",
    "from graph_functions import query_to_df,start_graph_session, start_graph_session_crashdown_w_auth\n",
    "from IPython.display import Image\n",
    "import numpy as np\n",
    "\n",
    "env = Env()\n",
    "#session = start_graph_session(env) \n",
    "session = start_graph_session_crashdown_w_auth(env) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Workflow\n",
    "### Initial exploration\n",
    "1. MR-eve results: find interesting exposure\n",
    "2. Manual MR: confirm the effect is there\n",
    "3. Broad review of literature triples connection (melodi-presto style)\n",
    "4. Traditional literature search with 'breast cancer'\n",
    "\n",
    "### Deeper dive into the role/mechanism\n",
    "\n",
    "5. Find related pathways (manual reactome search)\n",
    "    - compare with literature triples\n",
    "6. Pathway -> proteins -> genes\n",
    "7. GWAS -> tophits -> genes\n",
    "8. Compare gene overlap 6) and 7)\n",
    "\n",
    "9. Check tissue exprossion for genes from 7)\n",
    "\n",
    "### next part\n",
    "\n",
    "??\n",
    "link QTL?\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "MATCH (exposure:Gwas)-[mr:MR_EVE_MR]->(outcome:Gwas)\n",
      "WHERE outcome.id in ['ieu-a-1126','ieu-a-1127','ieu-a-1128']\n",
      "AND  exposure.id  = 'met-a-362'\n",
      "with mr, exposure, outcome\n",
      "ORDER BY exposure.trait\n",
      "RETURN exposure.id, exposure.trait, \n",
      "        outcome.id, \n",
      "        mr.pval, mr.b, mr.se, mr.method, mr.moescore\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exposure.id</th>\n",
       "      <th>exposure.trait</th>\n",
       "      <th>outcome.id</th>\n",
       "      <th>mr.pval</th>\n",
       "      <th>mr.b</th>\n",
       "      <th>mr.se</th>\n",
       "      <th>mr.method</th>\n",
       "      <th>mr.moescore</th>\n",
       "      <th>or</th>\n",
       "      <th>or_loci</th>\n",
       "      <th>or_upci</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>met-a-362</td>\n",
       "      <td>Betaine</td>\n",
       "      <td>ieu-a-1128</td>\n",
       "      <td>0.003013</td>\n",
       "      <td>-1.150799</td>\n",
       "      <td>0.387943</td>\n",
       "      <td>FE IVW</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.316384</td>\n",
       "      <td>0.147908</td>\n",
       "      <td>0.676765</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>met-a-362</td>\n",
       "      <td>Betaine</td>\n",
       "      <td>ieu-a-1127</td>\n",
       "      <td>0.300013</td>\n",
       "      <td>-0.173489</td>\n",
       "      <td>0.167395</td>\n",
       "      <td>FE IVW</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.840726</td>\n",
       "      <td>0.605571</td>\n",
       "      <td>1.167197</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>met-a-362</td>\n",
       "      <td>Betaine</td>\n",
       "      <td>ieu-a-1126</td>\n",
       "      <td>0.058986</td>\n",
       "      <td>-0.398726</td>\n",
       "      <td>0.211157</td>\n",
       "      <td>FE IVW</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.671174</td>\n",
       "      <td>0.443706</td>\n",
       "      <td>1.015256</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  exposure.id exposure.trait  outcome.id   mr.pval      mr.b     mr.se  \\\n",
       "0   met-a-362        Betaine  ieu-a-1128  0.003013 -1.150799  0.387943   \n",
       "1   met-a-362        Betaine  ieu-a-1127  0.300013 -0.173489  0.167395   \n",
       "2   met-a-362        Betaine  ieu-a-1126  0.058986 -0.398726  0.211157   \n",
       "\n",
       "  mr.method  mr.moescore        or   or_loci   or_upci  \n",
       "0    FE IVW          1.0  0.316384  0.147908  0.676765  \n",
       "1    FE IVW          1.0  0.840726  0.605571  1.167197  \n",
       "2    FE IVW          1.0  0.671174  0.443706  1.015256  "
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# MR results\n",
    "query = \"\"\"\n",
    "MATCH (exposure:Gwas)-[mr:MR_EVE_MR]->(outcome:Gwas)\n",
    "WHERE outcome.id in ['ieu-a-1126','ieu-a-1127','ieu-a-1128']\n",
    "AND  exposure.id  = 'met-a-362'\n",
    "with mr, exposure, outcome\n",
    "ORDER BY exposure.trait\n",
    "RETURN exposure.id, exposure.trait, \n",
    "        outcome.id, \n",
    "        mr.pval, mr.b, mr.se, mr.method, mr.moescore\n",
    "\"\"\"\n",
    "df = query_to_df(session, query, True)\n",
    "\n",
    "df['or'] = np.exp(df['mr.b'])\n",
    "df['or_loci'] = np.exp(df['mr.b'] - 1.96 * df['mr.se'])\n",
    "df['or_upci'] = np.exp(df['mr.b'] + 1.96 * df['mr.se'])\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# connection with literature\n",
    "\n",
    "# see R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "MATCH (g:Gwas)-[tophits:OPENGWAS_TOPHITS]->(v:Variant)-[]->(gene:Gene)\n",
      "where g.id = 'met-a-362'\n",
      "return g.id, v.name, gene.name\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>g.id</th>\n",
       "      <th>v.name</th>\n",
       "      <th>gene.name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>met-a-362</td>\n",
       "      <td>rs2851391</td>\n",
       "      <td>CBS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>met-a-362</td>\n",
       "      <td>rs2851391</td>\n",
       "      <td>DGKG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>met-a-362</td>\n",
       "      <td>rs2851391</td>\n",
       "      <td>ARHGEF11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>met-a-362</td>\n",
       "      <td>rs2851391</td>\n",
       "      <td>ZNF687</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>met-a-362</td>\n",
       "      <td>rs715</td>\n",
       "      <td>CPS1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>met-a-362</td>\n",
       "      <td>rs499368</td>\n",
       "      <td>SLC6A12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>met-a-362</td>\n",
       "      <td>rs499368</td>\n",
       "      <td>SLC6A12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>met-a-362</td>\n",
       "      <td>rs16876394</td>\n",
       "      <td>DMGDH</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        g.id      v.name gene.name\n",
       "0  met-a-362   rs2851391       CBS\n",
       "1  met-a-362   rs2851391      DGKG\n",
       "2  met-a-362   rs2851391  ARHGEF11\n",
       "3  met-a-362   rs2851391    ZNF687\n",
       "4  met-a-362       rs715      CPS1\n",
       "5  met-a-362    rs499368   SLC6A12\n",
       "6  met-a-362    rs499368   SLC6A12\n",
       "7  met-a-362  rs16876394     DMGDH"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# betaine GWAS -> pathways/expression data\n",
    "query = \"\"\"\n",
    "MATCH (g:Gwas)-[tophits:OPENGWAS_TOPHITS]->(v:Variant)-[]->(gene:Gene)\n",
    "where g.id = 'met-a-362'\n",
    "return g.id, v.name, gene.name\n",
    "\"\"\"\n",
    "df = query_to_df(session, query, True)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "MATCH (n:Pathway)-[]-(p:Protein)-[]-(g:Gene)\n",
      "where n.id in  ['R-HSA-6798163', 'R-HSA-442660']\n",
      "return n.name, p.name, g.name\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n.name</th>\n",
       "      <th>p.name</th>\n",
       "      <th>g.name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Choline catabolism</td>\n",
       "      <td>Q93088</td>\n",
       "      <td>BHMT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Choline catabolism</td>\n",
       "      <td>Q8WWI5</td>\n",
       "      <td>SLC44A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Choline catabolism</td>\n",
       "      <td>Q8NE62</td>\n",
       "      <td>CHDH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Choline catabolism</td>\n",
       "      <td>P49419</td>\n",
       "      <td>ALDH7A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Choline catabolism</td>\n",
       "      <td>Q9UL12</td>\n",
       "      <td>SARDH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Choline catabolism</td>\n",
       "      <td>Q9UI17</td>\n",
       "      <td>DMGDH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Na+/Cl- dependent neurotransmitter transporters</td>\n",
       "      <td>Q96N87</td>\n",
       "      <td>SLC6A18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Na+/Cl- dependent neurotransmitter transporters</td>\n",
       "      <td>Q695T7</td>\n",
       "      <td>SLC6A19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Na+/Cl- dependent neurotransmitter transporters</td>\n",
       "      <td>Q05940</td>\n",
       "      <td>SLC18A2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Na+/Cl- dependent neurotransmitter transporters</td>\n",
       "      <td>Q01959</td>\n",
       "      <td>SLC6A3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Na+/Cl- dependent neurotransmitter transporters</td>\n",
       "      <td>P54219</td>\n",
       "      <td>SLC18A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Na+/Cl- dependent neurotransmitter transporters</td>\n",
       "      <td>P48067</td>\n",
       "      <td>SLC6A9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Na+/Cl- dependent neurotransmitter transporters</td>\n",
       "      <td>P48066</td>\n",
       "      <td>SLC6A11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Na+/Cl- dependent neurotransmitter transporters</td>\n",
       "      <td>P48065</td>\n",
       "      <td>SLC6A12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Na+/Cl- dependent neurotransmitter transporters</td>\n",
       "      <td>P31641</td>\n",
       "      <td>SLC6A6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Na+/Cl- dependent neurotransmitter transporters</td>\n",
       "      <td>P30531</td>\n",
       "      <td>SLC6A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Na+/Cl- dependent neurotransmitter transporters</td>\n",
       "      <td>P23975</td>\n",
       "      <td>SLC6A2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Na+/Cl- dependent neurotransmitter transporters</td>\n",
       "      <td>Q9Y345</td>\n",
       "      <td>SLC6A5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Na+/Cl- dependent neurotransmitter transporters</td>\n",
       "      <td>Q9UN76</td>\n",
       "      <td>SLC6A14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Na+/Cl- dependent neurotransmitter transporters</td>\n",
       "      <td>Q9NSD5</td>\n",
       "      <td>SLC6A13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Na+/Cl- dependent neurotransmitter transporters</td>\n",
       "      <td>O15245</td>\n",
       "      <td>SLC22A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Na+/Cl- dependent neurotransmitter transporters</td>\n",
       "      <td>O15244</td>\n",
       "      <td>SLC22A2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Na+/Cl- dependent neurotransmitter transporters</td>\n",
       "      <td>Q9NP91</td>\n",
       "      <td>SLC6A20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Na+/Cl- dependent neurotransmitter transporters</td>\n",
       "      <td>Q9H2J7</td>\n",
       "      <td>SLC6A15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Na+/Cl- dependent neurotransmitter transporters</td>\n",
       "      <td>Q99884</td>\n",
       "      <td>SLC6A7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             n.name  p.name   g.name\n",
       "0                                Choline catabolism  Q93088     BHMT\n",
       "1                                Choline catabolism  Q8WWI5  SLC44A1\n",
       "2                                Choline catabolism  Q8NE62     CHDH\n",
       "3                                Choline catabolism  P49419  ALDH7A1\n",
       "4                                Choline catabolism  Q9UL12    SARDH\n",
       "5                                Choline catabolism  Q9UI17    DMGDH\n",
       "6   Na+/Cl- dependent neurotransmitter transporters  Q96N87  SLC6A18\n",
       "7   Na+/Cl- dependent neurotransmitter transporters  Q695T7  SLC6A19\n",
       "8   Na+/Cl- dependent neurotransmitter transporters  Q05940  SLC18A2\n",
       "9   Na+/Cl- dependent neurotransmitter transporters  Q01959   SLC6A3\n",
       "10  Na+/Cl- dependent neurotransmitter transporters  P54219  SLC18A1\n",
       "11  Na+/Cl- dependent neurotransmitter transporters  P48067   SLC6A9\n",
       "12  Na+/Cl- dependent neurotransmitter transporters  P48066  SLC6A11\n",
       "13  Na+/Cl- dependent neurotransmitter transporters  P48065  SLC6A12\n",
       "14  Na+/Cl- dependent neurotransmitter transporters  P31641   SLC6A6\n",
       "15  Na+/Cl- dependent neurotransmitter transporters  P30531   SLC6A1\n",
       "16  Na+/Cl- dependent neurotransmitter transporters  P23975   SLC6A2\n",
       "17  Na+/Cl- dependent neurotransmitter transporters  Q9Y345   SLC6A5\n",
       "18  Na+/Cl- dependent neurotransmitter transporters  Q9UN76  SLC6A14\n",
       "19  Na+/Cl- dependent neurotransmitter transporters  Q9NSD5  SLC6A13\n",
       "20  Na+/Cl- dependent neurotransmitter transporters  O15245  SLC22A1\n",
       "21  Na+/Cl- dependent neurotransmitter transporters  O15244  SLC22A2\n",
       "22  Na+/Cl- dependent neurotransmitter transporters  Q9NP91  SLC6A20\n",
       "23  Na+/Cl- dependent neurotransmitter transporters  Q9H2J7  SLC6A15\n",
       "24  Na+/Cl- dependent neurotransmitter transporters  Q99884   SLC6A7"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# https://reactome.org/content/detail/R-HSA-6798163\n",
    "query = \"\"\"\n",
    "MATCH (n:Pathway)-[]-(p:Protein)-[]-(g:Gene)\n",
    "where n.id in  ['R-HSA-6798163', 'R-HSA-442660']\n",
    "return n.name, p.name, g.name\n",
    "\"\"\"\n",
    "df = query_to_df(session, query, True)\n",
    "df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "MATCH (g:Gwas)-[tophits:OPENGWAS_TOPHITS]->(v:Variant)-[]->(gene:Gene)-[gene_to_tissue:EXPRESSED_IN]->(tissue:Tissue)\n",
      "where g.id = 'met-a-362'\n",
      "AND gene_to_tissue.tpm > 1\n",
      "WITH  v, gene_to_tissue, gene, tissue\n",
      "ORDER BY gene_to_tissue.tpm  desc, tissue.name\n",
      "RETURN gene.name, count(tissue.name) as N_tissues, \n",
      "collect(tissue.name) as Tissues, collect(gene_to_tissue.tpm) as TPM_vals\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene.name</th>\n",
       "      <th>N_tissues</th>\n",
       "      <th>Tissues</th>\n",
       "      <th>TPM_vals</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CPS1</td>\n",
       "      <td>39</td>\n",
       "      <td>[Liver, Small Intestine - Terminal Ileum, Test...</td>\n",
       "      <td>[352.4, 36.46, 12.21, 9.24, 5.7395, 4.402, 2.9...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CBS</td>\n",
       "      <td>51</td>\n",
       "      <td>[Liver, Brain - Cerebellum, Brain - Cerebellar...</td>\n",
       "      <td>[135.9, 105.2, 87.155, 82.42, 72.08, 59.47, 52...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DGKG</td>\n",
       "      <td>47</td>\n",
       "      <td>[Brain - Cerebellar Hemisphere, Brain - Cerebe...</td>\n",
       "      <td>[60.23, 56.79, 42.365, 34.22, 11.47, 10.23, 10...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ARHGEF11</td>\n",
       "      <td>52</td>\n",
       "      <td>[Testis, Brain - Cerebellum, Brain - Cerebella...</td>\n",
       "      <td>[52.3, 49.61, 45.035, 31.66, 30.69, 30.56, 27....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DMGDH</td>\n",
       "      <td>33</td>\n",
       "      <td>[Liver, Kidney - Cortex, Adrenal Gland, Nerve ...</td>\n",
       "      <td>[40.8, 17.9, 4.8125, 4.811, 4.5895, 3.8965, 3....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ZNF687</td>\n",
       "      <td>52</td>\n",
       "      <td>[Testis, Bladder, Cells - EBV-transformed lymp...</td>\n",
       "      <td>[28.23, 23.37, 23.37, 23.15, 22.845, 22.49, 22...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SLC6A12</td>\n",
       "      <td>44</td>\n",
       "      <td>[Kidney - Cortex, Kidney - Cortex, Brain - Put...</td>\n",
       "      <td>[25.4, 25.4, 18.53, 18.53, 16.57, 16.57, 16.24...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene.name  N_tissues                                            Tissues  \\\n",
       "0      CPS1         39  [Liver, Small Intestine - Terminal Ileum, Test...   \n",
       "1       CBS         51  [Liver, Brain - Cerebellum, Brain - Cerebellar...   \n",
       "2      DGKG         47  [Brain - Cerebellar Hemisphere, Brain - Cerebe...   \n",
       "3  ARHGEF11         52  [Testis, Brain - Cerebellum, Brain - Cerebella...   \n",
       "4     DMGDH         33  [Liver, Kidney - Cortex, Adrenal Gland, Nerve ...   \n",
       "5    ZNF687         52  [Testis, Bladder, Cells - EBV-transformed lymp...   \n",
       "6   SLC6A12         44  [Kidney - Cortex, Kidney - Cortex, Brain - Put...   \n",
       "\n",
       "                                            TPM_vals  \n",
       "0  [352.4, 36.46, 12.21, 9.24, 5.7395, 4.402, 2.9...  \n",
       "1  [135.9, 105.2, 87.155, 82.42, 72.08, 59.47, 52...  \n",
       "2  [60.23, 56.79, 42.365, 34.22, 11.47, 10.23, 10...  \n",
       "3  [52.3, 49.61, 45.035, 31.66, 30.69, 30.56, 27....  \n",
       "4  [40.8, 17.9, 4.8125, 4.811, 4.5895, 3.8965, 3....  \n",
       "5  [28.23, 23.37, 23.37, 23.15, 22.845, 22.49, 22...  \n",
       "6  [25.4, 25.4, 18.53, 18.53, 16.57, 16.57, 16.24...  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check tissue expression\n",
    "query = \"\"\"\n",
    "MATCH (g:Gwas)-[tophits:OPENGWAS_TOPHITS]->(v:Variant)-[]->(gene:Gene)-[gene_to_tissue:EXPRESSED_IN]->(tissue:Tissue)\n",
    "where g.id = 'met-a-362'\n",
    "AND gene_to_tissue.tpm > 1\n",
    "WITH  v, gene_to_tissue, gene, tissue\n",
    "ORDER BY gene_to_tissue.tpm  desc, tissue.name\n",
    "RETURN gene.name, count(tissue.name) as N_tissues, \n",
    "collect(tissue.name) as Tissues, collect(gene_to_tissue.tpm) as TPM_vals\n",
    "\"\"\"\n",
    "df = query_to_df(session, query, True)\n",
    "df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "MATCH (g:Gwas)-[tophits:OPENGWAS_TOPHITS]->(v:Variant)-[]->(gene:Gene)-[gene_to_tissue:EXPRESSED_IN]->(tissue:Tissue)\n",
      "where g.id = 'met-a-362'\n",
      "AND  tissue.name = \"Breast - Mammary Tissue\"\n",
      "return distinct collect(v.name) as variants,  g.id,  gene.name,   gene_to_tissue.tpm as TPM, tissue.name\n",
      "ORDER BY TPM desc\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variants</th>\n",
       "      <th>g.id</th>\n",
       "      <th>gene.name</th>\n",
       "      <th>TPM</th>\n",
       "      <th>tissue.name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>[rs2851391]</td>\n",
       "      <td>met-a-362</td>\n",
       "      <td>ARHGEF11</td>\n",
       "      <td>18.85500</td>\n",
       "      <td>Breast - Mammary Tissue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>[rs2851391]</td>\n",
       "      <td>met-a-362</td>\n",
       "      <td>ZNF687</td>\n",
       "      <td>14.26000</td>\n",
       "      <td>Breast - Mammary Tissue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>[rs2851391]</td>\n",
       "      <td>met-a-362</td>\n",
       "      <td>CBS</td>\n",
       "      <td>6.28350</td>\n",
       "      <td>Breast - Mammary Tissue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>[rs16876394]</td>\n",
       "      <td>met-a-362</td>\n",
       "      <td>DMGDH</td>\n",
       "      <td>2.46300</td>\n",
       "      <td>Breast - Mammary Tissue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>[rs715]</td>\n",
       "      <td>met-a-362</td>\n",
       "      <td>CPS1</td>\n",
       "      <td>2.10150</td>\n",
       "      <td>Breast - Mammary Tissue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>[rs2851391]</td>\n",
       "      <td>met-a-362</td>\n",
       "      <td>DGKG</td>\n",
       "      <td>1.85550</td>\n",
       "      <td>Breast - Mammary Tissue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>[rs499368, rs499368]</td>\n",
       "      <td>met-a-362</td>\n",
       "      <td>SLC6A12</td>\n",
       "      <td>0.38425</td>\n",
       "      <td>Breast - Mammary Tissue</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               variants       g.id gene.name       TPM  \\\n",
       "0           [rs2851391]  met-a-362  ARHGEF11  18.85500   \n",
       "1           [rs2851391]  met-a-362    ZNF687  14.26000   \n",
       "2           [rs2851391]  met-a-362       CBS   6.28350   \n",
       "3          [rs16876394]  met-a-362     DMGDH   2.46300   \n",
       "4               [rs715]  met-a-362      CPS1   2.10150   \n",
       "5           [rs2851391]  met-a-362      DGKG   1.85550   \n",
       "6  [rs499368, rs499368]  met-a-362   SLC6A12   0.38425   \n",
       "\n",
       "               tissue.name  \n",
       "0  Breast - Mammary Tissue  \n",
       "1  Breast - Mammary Tissue  \n",
       "2  Breast - Mammary Tissue  \n",
       "3  Breast - Mammary Tissue  \n",
       "4  Breast - Mammary Tissue  \n",
       "5  Breast - Mammary Tissue  \n",
       "6  Breast - Mammary Tissue  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#  expression in breast tissue (probably does not make sense? )\n",
    "query = \"\"\"\n",
    "MATCH (g:Gwas)-[tophits:OPENGWAS_TOPHITS]->(v:Variant)-[]->(gene:Gene)-[gene_to_tissue:EXPRESSED_IN]->(tissue:Tissue)\n",
    "where g.id = 'met-a-362'\n",
    "AND  tissue.name = \"Breast - Mammary Tissue\"\n",
    "return distinct collect(v.name) as variants,  g.id,  gene.name,   gene_to_tissue.tpm as TPM, tissue.name\n",
    "ORDER BY TPM desc\n",
    "\"\"\"\n",
    "df = query_to_df(session, query, True)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "MATCH (n:Pathway)-[]-(p:Protein)-[]-(g:Gene)-[gene_to_tissue:EXPRESSED_IN]->(tissue:Tissue)\n",
      "where n.id = 'R-HSA-6798163' //chlorine catabolism\n",
      "AND gene_to_tissue.tpm > 1\n",
      "WITH   gene_to_tissue, g, tissue\n",
      "ORDER BY gene_to_tissue.tpm  desc, tissue.name\n",
      "RETURN g.name, count(tissue.name) as N_tissues, \n",
      "collect(tissue.name) as Tissues, collect(gene_to_tissue.tpm) as TPM_vals\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>g.name</th>\n",
       "      <th>N_tissues</th>\n",
       "      <th>Tissues</th>\n",
       "      <th>TPM_vals</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BHMT</td>\n",
       "      <td>9</td>\n",
       "      <td>[Liver, Kidney - Cortex, Bladder, Esophagus - ...</td>\n",
       "      <td>[287.6, 105.0, 18.46, 4.9030000000000005, 2.43...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SLC44A1</td>\n",
       "      <td>52</td>\n",
       "      <td>[Brain - Spinal cord (cervical c-1), Brain - S...</td>\n",
       "      <td>[109.2, 34.925, 30.755, 29.065, 28.08, 25.995,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ALDH7A1</td>\n",
       "      <td>51</td>\n",
       "      <td>[Ovary, Brain - Amygdala, Brain - Caudate (bas...</td>\n",
       "      <td>[60.24, 53.625, 51.32, 45.83, 41.31, 40.5, 40....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SARDH</td>\n",
       "      <td>32</td>\n",
       "      <td>[Liver, Pancreas, Kidney - Cortex, Pituitary, ...</td>\n",
       "      <td>[41.64, 11.7, 3.7880000000000003, 3.728, 3.66,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DMGDH</td>\n",
       "      <td>33</td>\n",
       "      <td>[Liver, Kidney - Cortex, Adrenal Gland, Nerve ...</td>\n",
       "      <td>[40.8, 17.9, 4.8125, 4.811, 4.5895, 3.8965, 3....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CHDH</td>\n",
       "      <td>45</td>\n",
       "      <td>[Liver, Kidney - Cortex, Minor Salivary Gland,...</td>\n",
       "      <td>[25.77, 19.99, 18.84, 17.58, 17.04, 16.9, 16.5...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    g.name  N_tissues                                            Tissues  \\\n",
       "0     BHMT          9  [Liver, Kidney - Cortex, Bladder, Esophagus - ...   \n",
       "1  SLC44A1         52  [Brain - Spinal cord (cervical c-1), Brain - S...   \n",
       "2  ALDH7A1         51  [Ovary, Brain - Amygdala, Brain - Caudate (bas...   \n",
       "3    SARDH         32  [Liver, Pancreas, Kidney - Cortex, Pituitary, ...   \n",
       "4    DMGDH         33  [Liver, Kidney - Cortex, Adrenal Gland, Nerve ...   \n",
       "5     CHDH         45  [Liver, Kidney - Cortex, Minor Salivary Gland,...   \n",
       "\n",
       "                                            TPM_vals  \n",
       "0  [287.6, 105.0, 18.46, 4.9030000000000005, 2.43...  \n",
       "1  [109.2, 34.925, 30.755, 29.065, 28.08, 25.995,...  \n",
       "2  [60.24, 53.625, 51.32, 45.83, 41.31, 40.5, 40....  \n",
       "3  [41.64, 11.7, 3.7880000000000003, 3.728, 3.66,...  \n",
       "4  [40.8, 17.9, 4.8125, 4.811, 4.5895, 3.8965, 3....  \n",
       "5  [25.77, 19.99, 18.84, 17.58, 17.04, 16.9, 16.5...  "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"\"\"\n",
    "MATCH (n:Pathway)-[]-(p:Protein)-[]-(g:Gene)-[gene_to_tissue:EXPRESSED_IN]->(tissue:Tissue)\n",
    "where n.id = 'R-HSA-6798163' //chlorine catabolism\n",
    "AND gene_to_tissue.tpm > 1\n",
    "WITH   gene_to_tissue, g, tissue\n",
    "ORDER BY gene_to_tissue.tpm  desc, tissue.name\n",
    "RETURN g.name, count(tissue.name) as N_tissues, \n",
    "collect(tissue.name) as Tissues, collect(gene_to_tissue.tpm) as TPM_vals\n",
    "\"\"\"\n",
    "df = query_to_df(session, query, True)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "MATCH (g:Gwas)-[tophits:OPENGWAS_TOPHITS]->(v:Variant)-[]->(gene:Gene)-[]-(prot:Protein)-[inpath:PROTEIN_IN_PATHWAY]-(path:Pathway)\n",
      "where g.id = 'met-a-362'\n",
      "return distinct collect(gene.name) as genes,  path.name\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genes</th>\n",
       "      <th>path.name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>[DGKG]</td>\n",
       "      <td>Effects of PIP2 hydrolysis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>[DGKG]</td>\n",
       "      <td>G alpha (q) signalling events</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>[DGKG, ARHGEF11]</td>\n",
       "      <td>GPCR downstream signalling</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>[DGKG]</td>\n",
       "      <td>Hemostasis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>[DGKG]</td>\n",
       "      <td>Platelet activation, signaling and aggregation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>[DGKG, ARHGEF11]</td>\n",
       "      <td>Signal Transduction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>[DGKG, ARHGEF11]</td>\n",
       "      <td>Signaling by GPCR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>[SLC6A12, SLC6A12]</td>\n",
       "      <td>Amino acid transport across the plasma membrane</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>[SLC6A12, SLC6A12]</td>\n",
       "      <td>Creatine metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>[SLC6A12, SLC6A12]</td>\n",
       "      <td>GABA synthesis, release, reuptake and degradation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>[SLC6A12, SLC6A12, CBS, CPS1, DMGDH]</td>\n",
       "      <td>Metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>[SLC6A12, SLC6A12, CBS, CPS1, DMGDH]</td>\n",
       "      <td>Metabolism of amino acids and derivatives</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>[SLC6A12, SLC6A12]</td>\n",
       "      <td>Na+/Cl- dependent neurotransmitter transporters</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>[SLC6A12, SLC6A12]</td>\n",
       "      <td>Neuronal System</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>[SLC6A12, SLC6A12]</td>\n",
       "      <td>Neurotransmitter release cycle</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>[SLC6A12, SLC6A12]</td>\n",
       "      <td>Reuptake of GABA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>[SLC6A12, SLC6A12]</td>\n",
       "      <td>SLC-mediated transmembrane transport</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>[SLC6A12, SLC6A12]</td>\n",
       "      <td>Transmission across Chemical Synapses</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>[SLC6A12, SLC6A12]</td>\n",
       "      <td>Transport of bile salts and organic acids, met...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>[SLC6A12, SLC6A12]</td>\n",
       "      <td>Transport of inorganic cations/anions and amin...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>[SLC6A12, SLC6A12]</td>\n",
       "      <td>Transport of small molecules</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>[CBS]</td>\n",
       "      <td>Cysteine formation from homocysteine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>[CBS]</td>\n",
       "      <td>Metabolism of ingested SeMet, Sec, MeSec into ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>[CBS]</td>\n",
       "      <td>Selenoamino acid metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>[CBS]</td>\n",
       "      <td>Sulfur amino acid metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>[CPS1]</td>\n",
       "      <td>Urea cycle</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>[DMGDH]</td>\n",
       "      <td>Choline catabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>[ARHGEF11]</td>\n",
       "      <td>Axon guidance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>[ARHGEF11]</td>\n",
       "      <td>Cell death signalling via NRAGE, NRIF and NADE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>[ARHGEF11]</td>\n",
       "      <td>Death Receptor Signalling</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>[ARHGEF11]</td>\n",
       "      <td>Developmental Biology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>[ARHGEF11]</td>\n",
       "      <td>G alpha (12/13) signalling events</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>[ARHGEF11]</td>\n",
       "      <td>NRAGE signals death through JNK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>[ARHGEF11]</td>\n",
       "      <td>Nervous system development</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>[ARHGEF11]</td>\n",
       "      <td>Sema4D in semaphorin signaling</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>[ARHGEF11]</td>\n",
       "      <td>Sema4D induced cell migration and growth-cone ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>[ARHGEF11]</td>\n",
       "      <td>Semaphorin interactions</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>[ARHGEF11]</td>\n",
       "      <td>Signaling by Rho GTPases</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>[ARHGEF11]</td>\n",
       "      <td>p75 NTR receptor-mediated signalling</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   genes  \\\n",
       "0                                 [DGKG]   \n",
       "1                                 [DGKG]   \n",
       "2                       [DGKG, ARHGEF11]   \n",
       "3                                 [DGKG]   \n",
       "4                                 [DGKG]   \n",
       "5                       [DGKG, ARHGEF11]   \n",
       "6                       [DGKG, ARHGEF11]   \n",
       "7                     [SLC6A12, SLC6A12]   \n",
       "8                     [SLC6A12, SLC6A12]   \n",
       "9                     [SLC6A12, SLC6A12]   \n",
       "10  [SLC6A12, SLC6A12, CBS, CPS1, DMGDH]   \n",
       "11  [SLC6A12, SLC6A12, CBS, CPS1, DMGDH]   \n",
       "12                    [SLC6A12, SLC6A12]   \n",
       "13                    [SLC6A12, SLC6A12]   \n",
       "14                    [SLC6A12, SLC6A12]   \n",
       "15                    [SLC6A12, SLC6A12]   \n",
       "16                    [SLC6A12, SLC6A12]   \n",
       "17                    [SLC6A12, SLC6A12]   \n",
       "18                    [SLC6A12, SLC6A12]   \n",
       "19                    [SLC6A12, SLC6A12]   \n",
       "20                    [SLC6A12, SLC6A12]   \n",
       "21                                 [CBS]   \n",
       "22                                 [CBS]   \n",
       "23                                 [CBS]   \n",
       "24                                 [CBS]   \n",
       "25                                [CPS1]   \n",
       "26                               [DMGDH]   \n",
       "27                            [ARHGEF11]   \n",
       "28                            [ARHGEF11]   \n",
       "29                            [ARHGEF11]   \n",
       "30                            [ARHGEF11]   \n",
       "31                            [ARHGEF11]   \n",
       "32                            [ARHGEF11]   \n",
       "33                            [ARHGEF11]   \n",
       "34                            [ARHGEF11]   \n",
       "35                            [ARHGEF11]   \n",
       "36                            [ARHGEF11]   \n",
       "37                            [ARHGEF11]   \n",
       "38                            [ARHGEF11]   \n",
       "\n",
       "                                            path.name  \n",
       "0                          Effects of PIP2 hydrolysis  \n",
       "1                       G alpha (q) signalling events  \n",
       "2                          GPCR downstream signalling  \n",
       "3                                          Hemostasis  \n",
       "4      Platelet activation, signaling and aggregation  \n",
       "5                                 Signal Transduction  \n",
       "6                                   Signaling by GPCR  \n",
       "7     Amino acid transport across the plasma membrane  \n",
       "8                                 Creatine metabolism  \n",
       "9   GABA synthesis, release, reuptake and degradation  \n",
       "10                                         Metabolism  \n",
       "11          Metabolism of amino acids and derivatives  \n",
       "12    Na+/Cl- dependent neurotransmitter transporters  \n",
       "13                                    Neuronal System  \n",
       "14                     Neurotransmitter release cycle  \n",
       "15                                   Reuptake of GABA  \n",
       "16               SLC-mediated transmembrane transport  \n",
       "17              Transmission across Chemical Synapses  \n",
       "18  Transport of bile salts and organic acids, met...  \n",
       "19  Transport of inorganic cations/anions and amin...  \n",
       "20                       Transport of small molecules  \n",
       "21               Cysteine formation from homocysteine  \n",
       "22  Metabolism of ingested SeMet, Sec, MeSec into ...  \n",
       "23                        Selenoamino acid metabolism  \n",
       "24                       Sulfur amino acid metabolism  \n",
       "25                                         Urea cycle  \n",
       "26                                 Choline catabolism  \n",
       "27                                      Axon guidance  \n",
       "28     Cell death signalling via NRAGE, NRIF and NADE  \n",
       "29                          Death Receptor Signalling  \n",
       "30                              Developmental Biology  \n",
       "31                  G alpha (12/13) signalling events  \n",
       "32                    NRAGE signals death through JNK  \n",
       "33                         Nervous system development  \n",
       "34                     Sema4D in semaphorin signaling  \n",
       "35  Sema4D induced cell migration and growth-cone ...  \n",
       "36                            Semaphorin interactions  \n",
       "37                           Signaling by Rho GTPases  \n",
       "38               p75 NTR receptor-mediated signalling  "
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# other pathways that genes from gwas are involved in?\n",
    "query = \"\"\"\n",
    "MATCH (g:Gwas)-[tophits:OPENGWAS_TOPHITS]->(v:Variant)-[]->(gene:Gene)-[]-(prot:Protein)-[inpath:PROTEIN_IN_PATHWAY]-(path:Pathway)\n",
    "where g.id = 'met-a-362'\n",
    "return distinct collect(gene.name) as genes,  path.name\n",
    "\"\"\"\n",
    "df = query_to_df(session, query, True)\n",
    "df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
